{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import pathlib \n",
    "import optuna\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import random\n",
    "import torch\n",
    "import sys\n",
    "import blitzgsea as blitz\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "from sklearn.decomposition import PCA, FastICA, NMF\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "from torch.utils.data import DataLoader, TensorDataset\n",
    "from scipy.stats import ttest_ind\n",
    "\n",
    "script_directory = pathlib.Path(\"../2.train-VAE/utils/\").resolve()\n",
    "sys.path.insert(0, str(script_directory))\n",
    "from betavae import BetaVAE, train_vae, weights\n",
    "from betatcvae import BetaTCVAE, train_tc_vae, tc_weights\n",
    "from vanillavae import VanillaVAE, train_vvae, vanilla_weights\n",
    "from optimize_utils import get_optimize_args, objective, get_optimizer\n",
    "from optimize_utils_tcvae import get_optimize_args_tc, objective_tc, get_optimizer_tc\n",
    "from optimize_utils_vvae import get_optimize_args_vvae, objective_vvae, get_optimizer_vvae\n",
    "\n",
    "script_directory = pathlib.Path(\"../utils/\").resolve()\n",
    "sys.path.insert(0, str(script_directory))\n",
    "from data_loader import load_train_test_data, load_model_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1150, 18444)\n"
     ]
    }
   ],
   "source": [
    "# Load command line arguments\n",
    "args = get_optimize_args()\n",
    "tc_args = get_optimize_args_tc()\n",
    "vvae_args = get_optimize_args_vvae()\n",
    "\n",
    "# Load data\n",
    "data_directory = pathlib.Path(\"../0.data-download/data\").resolve()\n",
    "\n",
    "train_df, test_df, val_df, load_gene_stats = load_train_test_data(\n",
    "    data_directory, train_or_test=\"all\", load_gene_stats=True, zero_one_normalize=True\n",
    ")\n",
    "train_data = pd.DataFrame(train_df)\n",
    "\n",
    "dependency_file = pathlib.Path(f\"{data_directory}/CRISPRGeneEffect.parquet\").resolve()\n",
    "gene_dict_file = pathlib.Path(f\"{data_directory}/CRISPR_gene_dictionary.parquet\").resolve()\n",
    "dependency_df, gene_dict_df= load_model_data(dependency_file, gene_dict_file)\n",
    "gene_dict_df = pd.DataFrame(gene_dict_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FDX2 (112812)</th>\n",
       "      <th>NSMCE4A (54780)</th>\n",
       "      <th>AASDHPPT (60496)</th>\n",
       "      <th>RAD50 (10111)</th>\n",
       "      <th>PRMT5 (10419)</th>\n",
       "      <th>ALG6 (29929)</th>\n",
       "      <th>EXOC1 (55763)</th>\n",
       "      <th>RHNO1 (83695)</th>\n",
       "      <th>ATG12 (9140)</th>\n",
       "      <th>ITPK1 (3705)</th>\n",
       "      <th>...</th>\n",
       "      <th>CDC25B (994)</th>\n",
       "      <th>CNOT4 (4850)</th>\n",
       "      <th>WDR83 (84292)</th>\n",
       "      <th>UBALD2 (283991)</th>\n",
       "      <th>DOT1L (84444)</th>\n",
       "      <th>DNM1L (10059)</th>\n",
       "      <th>PET117 (100303755)</th>\n",
       "      <th>RNF168 (165918)</th>\n",
       "      <th>SENP8 (123228)</th>\n",
       "      <th>SECISBP2 (79048)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-0.880565</td>\n",
       "      <td>-0.782422</td>\n",
       "      <td>-0.778251</td>\n",
       "      <td>-0.187087</td>\n",
       "      <td>-1.177537</td>\n",
       "      <td>-0.213722</td>\n",
       "      <td>-0.403387</td>\n",
       "      <td>-0.028371</td>\n",
       "      <td>0.043060</td>\n",
       "      <td>-0.078512</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.172492</td>\n",
       "      <td>0.196713</td>\n",
       "      <td>-0.185486</td>\n",
       "      <td>-0.357539</td>\n",
       "      <td>-0.062151</td>\n",
       "      <td>-0.181394</td>\n",
       "      <td>-0.100426</td>\n",
       "      <td>-0.425072</td>\n",
       "      <td>0.003515</td>\n",
       "      <td>-0.336083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-0.749120</td>\n",
       "      <td>-0.516381</td>\n",
       "      <td>-0.080714</td>\n",
       "      <td>0.056364</td>\n",
       "      <td>-0.466129</td>\n",
       "      <td>-0.245435</td>\n",
       "      <td>-0.899518</td>\n",
       "      <td>-0.135167</td>\n",
       "      <td>-0.259756</td>\n",
       "      <td>-0.167318</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.256712</td>\n",
       "      <td>0.465617</td>\n",
       "      <td>-0.284650</td>\n",
       "      <td>-0.203330</td>\n",
       "      <td>-0.107819</td>\n",
       "      <td>-0.684306</td>\n",
       "      <td>0.141742</td>\n",
       "      <td>-0.104726</td>\n",
       "      <td>-0.087244</td>\n",
       "      <td>-0.022045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-0.575482</td>\n",
       "      <td>-0.353394</td>\n",
       "      <td>-0.083243</td>\n",
       "      <td>0.076057</td>\n",
       "      <td>-0.953753</td>\n",
       "      <td>-0.251060</td>\n",
       "      <td>-0.902062</td>\n",
       "      <td>-0.093195</td>\n",
       "      <td>-0.043493</td>\n",
       "      <td>-0.141663</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.446836</td>\n",
       "      <td>0.054560</td>\n",
       "      <td>-0.494846</td>\n",
       "      <td>0.071733</td>\n",
       "      <td>0.258250</td>\n",
       "      <td>-0.370277</td>\n",
       "      <td>-0.059265</td>\n",
       "      <td>-0.713488</td>\n",
       "      <td>0.090763</td>\n",
       "      <td>-0.299328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-0.475950</td>\n",
       "      <td>-0.310926</td>\n",
       "      <td>0.048373</td>\n",
       "      <td>0.064619</td>\n",
       "      <td>-1.014308</td>\n",
       "      <td>-0.407698</td>\n",
       "      <td>-0.765044</td>\n",
       "      <td>-0.196072</td>\n",
       "      <td>0.059620</td>\n",
       "      <td>-0.307384</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.229200</td>\n",
       "      <td>-0.077849</td>\n",
       "      <td>0.027445</td>\n",
       "      <td>-0.074084</td>\n",
       "      <td>-0.541231</td>\n",
       "      <td>-0.420726</td>\n",
       "      <td>-0.092938</td>\n",
       "      <td>-1.170949</td>\n",
       "      <td>0.135269</td>\n",
       "      <td>-0.194019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-0.604045</td>\n",
       "      <td>-0.210099</td>\n",
       "      <td>-0.567472</td>\n",
       "      <td>-0.106073</td>\n",
       "      <td>-0.733926</td>\n",
       "      <td>-0.223492</td>\n",
       "      <td>-0.325947</td>\n",
       "      <td>-0.015823</td>\n",
       "      <td>0.244816</td>\n",
       "      <td>-0.422609</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.049858</td>\n",
       "      <td>0.202080</td>\n",
       "      <td>-0.217986</td>\n",
       "      <td>-0.402349</td>\n",
       "      <td>-0.482002</td>\n",
       "      <td>-0.851300</td>\n",
       "      <td>-0.654365</td>\n",
       "      <td>-0.428541</td>\n",
       "      <td>0.077339</td>\n",
       "      <td>0.096456</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 2718 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   FDX2 (112812)  NSMCE4A (54780)  AASDHPPT (60496)  RAD50 (10111)  \\\n",
       "0      -0.880565        -0.782422         -0.778251      -0.187087   \n",
       "1      -0.749120        -0.516381         -0.080714       0.056364   \n",
       "2      -0.575482        -0.353394         -0.083243       0.076057   \n",
       "3      -0.475950        -0.310926          0.048373       0.064619   \n",
       "4      -0.604045        -0.210099         -0.567472      -0.106073   \n",
       "\n",
       "   PRMT5 (10419)  ALG6 (29929)  EXOC1 (55763)  RHNO1 (83695)  ATG12 (9140)  \\\n",
       "0      -1.177537     -0.213722      -0.403387      -0.028371      0.043060   \n",
       "1      -0.466129     -0.245435      -0.899518      -0.135167     -0.259756   \n",
       "2      -0.953753     -0.251060      -0.902062      -0.093195     -0.043493   \n",
       "3      -1.014308     -0.407698      -0.765044      -0.196072      0.059620   \n",
       "4      -0.733926     -0.223492      -0.325947      -0.015823      0.244816   \n",
       "\n",
       "   ITPK1 (3705)  ...  CDC25B (994)  CNOT4 (4850)  WDR83 (84292)  \\\n",
       "0     -0.078512  ...     -0.172492      0.196713      -0.185486   \n",
       "1     -0.167318  ...     -0.256712      0.465617      -0.284650   \n",
       "2     -0.141663  ...     -0.446836      0.054560      -0.494846   \n",
       "3     -0.307384  ...     -0.229200     -0.077849       0.027445   \n",
       "4     -0.422609  ...     -0.049858      0.202080      -0.217986   \n",
       "\n",
       "   UBALD2 (283991)  DOT1L (84444)  DNM1L (10059)  PET117 (100303755)  \\\n",
       "0        -0.357539      -0.062151      -0.181394           -0.100426   \n",
       "1        -0.203330      -0.107819      -0.684306            0.141742   \n",
       "2         0.071733       0.258250      -0.370277           -0.059265   \n",
       "3        -0.074084      -0.541231      -0.420726           -0.092938   \n",
       "4        -0.402349      -0.482002      -0.851300           -0.654365   \n",
       "\n",
       "   RNF168 (165918)  SENP8 (123228)  SECISBP2 (79048)  \n",
       "0        -0.425072        0.003515         -0.336083  \n",
       "1        -0.104726       -0.087244         -0.022045  \n",
       "2        -0.713488        0.090763         -0.299328  \n",
       "3        -1.170949        0.135269         -0.194019  \n",
       "4        -0.428541        0.077339          0.096456  \n",
       "\n",
       "[5 rows x 2718 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Load weight data for VAEs\n",
    "data_directory = pathlib.Path(\"../0.data-download/data\").resolve()\n",
    "weight_df = load_train_test_data(\n",
    "    data_directory, train_or_test=\"train\"\n",
    ")\n",
    "\n",
    "gene_list_passed_qc = gene_dict_df.loc[\n",
    "    gene_dict_df[\"qc_pass\"], \"dependency_column\"\n",
    "].tolist()\n",
    "\n",
    "weight_data = weight_df.filter(gene_list_passed_qc, axis=1)\n",
    "weight_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert dataframes to tensors\n",
    "train_tensor = torch.tensor(train_df, dtype=torch.float32)\n",
    "test_tensor = torch.tensor(test_df, dtype=torch.float32)\n",
    "val_tensor = torch.tensor(val_df, dtype=torch.float32)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_dir = pathlib.Path(\"./results\").resolve()\n",
    "output_dir.mkdir(parents=True, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to perform dimensionality reduction and extract z matrices for sklearn models\n",
    "def extract_z_matrix_sklearn(model, data, model_name, num_components, output_dir):\n",
    "    if model_name == \"nmf\":\n",
    "        min_value = data.min().min()\n",
    "        if min_value < 0:\n",
    "            data = data - min_value\n",
    "    # Fit the model and transform the data\n",
    "    z_matrix = model.fit_transform(data)\n",
    "    \n",
    "    # Create a DataFrame and save it\n",
    "    z_matrix_df = pd.DataFrame(z_matrix, columns=[f'{model_name}_{i}' for i in range(num_components)])\n",
    "    z_matrix_df.insert(0, 'ModelID', train_data.index) \n",
    "    \n",
    "    return z_matrix_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded existing combined results from /home/juliacurd/gene_dependency_representations/4.gene_expression_signatures/results/combined_z_matrix_gsea_results.parquet\n",
      "Skipping pca with 2 dimensions as it is already processed.\n",
      "Skipping ica with 2 dimensions as it is already processed.\n",
      "Skipping nmf with 2 dimensions as it is already processed.\n",
      "Skipping vanillavae with 2 dimensions as it is already processed.\n",
      "Skipping betavae with 2 dimensions as it is already processed.\n",
      "Skipping betatcvae with 2 dimensions as it is already processed.\n",
      "Skipping pca with 3 dimensions as it is already processed.\n",
      "Skipping ica with 3 dimensions as it is already processed.\n",
      "Skipping nmf with 3 dimensions as it is already processed.\n",
      "Skipping vanillavae with 3 dimensions as it is already processed.\n",
      "Skipping betavae with 3 dimensions as it is already processed.\n",
      "Skipping betatcvae with 3 dimensions as it is already processed.\n",
      "Skipping pca with 4 dimensions as it is already processed.\n",
      "Skipping ica with 4 dimensions as it is already processed.\n",
      "Skipping nmf with 4 dimensions as it is already processed.\n",
      "Skipping vanillavae with 4 dimensions as it is already processed.\n",
      "Skipping betavae with 4 dimensions as it is already processed.\n",
      "Skipping betatcvae with 4 dimensions as it is already processed.\n",
      "Skipping pca with 5 dimensions as it is already processed.\n",
      "Skipping ica with 5 dimensions as it is already processed.\n",
      "Skipping nmf with 5 dimensions as it is already processed.\n",
      "Skipping vanillavae with 5 dimensions as it is already processed.\n",
      "Skipping betavae with 5 dimensions as it is already processed.\n",
      "Skipping betatcvae with 5 dimensions as it is already processed.\n",
      "Skipping pca with 6 dimensions as it is already processed.\n",
      "Skipping ica with 6 dimensions as it is already processed.\n",
      "Skipping nmf with 6 dimensions as it is already processed.\n",
      "Skipping vanillavae with 6 dimensions as it is already processed.\n",
      "Skipping betavae with 6 dimensions as it is already processed.\n",
      "Skipping betatcvae with 6 dimensions as it is already processed.\n",
      "Skipping pca with 7 dimensions as it is already processed.\n",
      "Skipping ica with 7 dimensions as it is already processed.\n",
      "Skipping nmf with 7 dimensions as it is already processed.\n",
      "Skipping vanillavae with 7 dimensions as it is already processed.\n",
      "Skipping betavae with 7 dimensions as it is already processed.\n",
      "Skipping betatcvae with 7 dimensions as it is already processed.\n",
      "Skipping pca with 8 dimensions as it is already processed.\n",
      "Skipping ica with 8 dimensions as it is already processed.\n",
      "Skipping nmf with 8 dimensions as it is already processed.\n",
      "Skipping vanillavae with 8 dimensions as it is already processed.\n",
      "Skipping betavae with 8 dimensions as it is already processed.\n",
      "Skipping betatcvae with 8 dimensions as it is already processed.\n",
      "Skipping pca with 9 dimensions as it is already processed.\n",
      "Skipping ica with 9 dimensions as it is already processed.\n",
      "Skipping nmf with 9 dimensions as it is already processed.\n",
      "Skipping vanillavae with 9 dimensions as it is already processed.\n",
      "Skipping betavae with 9 dimensions as it is already processed.\n",
      "Skipping betatcvae with 9 dimensions as it is already processed.\n",
      "Skipping pca with 10 dimensions as it is already processed.\n",
      "Skipping ica with 10 dimensions as it is already processed.\n",
      "Skipping nmf with 10 dimensions as it is already processed.\n",
      "Skipping vanillavae with 10 dimensions as it is already processed.\n",
      "Skipping betavae with 10 dimensions as it is already processed.\n",
      "Skipping betatcvae with 10 dimensions as it is already processed.\n",
      "Skipping pca with 12 dimensions as it is already processed.\n",
      "Skipping ica with 12 dimensions as it is already processed.\n",
      "Skipping nmf with 12 dimensions as it is already processed.\n",
      "Skipping vanillavae with 12 dimensions as it is already processed.\n",
      "Skipping betavae with 12 dimensions as it is already processed.\n",
      "Skipping betatcvae with 12 dimensions as it is already processed.\n",
      "Skipping pca with 14 dimensions as it is already processed.\n",
      "Skipping ica with 14 dimensions as it is already processed.\n",
      "Skipping nmf with 14 dimensions as it is already processed.\n",
      "Skipping vanillavae with 14 dimensions as it is already processed.\n",
      "Skipping betavae with 14 dimensions as it is already processed.\n",
      "Skipping betatcvae with 14 dimensions as it is already processed.\n",
      "Skipping pca with 16 dimensions as it is already processed.\n",
      "Skipping ica with 16 dimensions as it is already processed.\n",
      "Skipping nmf with 16 dimensions as it is already processed.\n",
      "Skipping vanillavae with 16 dimensions as it is already processed.\n",
      "Skipping betavae with 16 dimensions as it is already processed.\n",
      "Skipping betatcvae with 16 dimensions as it is already processed.\n",
      "Skipping pca with 18 dimensions as it is already processed.\n",
      "Skipping ica with 18 dimensions as it is already processed.\n",
      "Skipping nmf with 18 dimensions as it is already processed.\n",
      "Skipping vanillavae with 18 dimensions as it is already processed.\n",
      "Skipping betavae with 18 dimensions as it is already processed.\n",
      "Skipping betatcvae with 18 dimensions as it is already processed.\n",
      "Skipping pca with 20 dimensions as it is already processed.\n",
      "Skipping ica with 20 dimensions as it is already processed.\n",
      "Skipping nmf with 20 dimensions as it is already processed.\n",
      "Skipping vanillavae with 20 dimensions as it is already processed.\n",
      "Skipping betavae with 20 dimensions as it is already processed.\n",
      "Skipping betatcvae with 20 dimensions as it is already processed.\n",
      "Skipping pca with 25 dimensions as it is already processed.\n",
      "Skipping ica with 25 dimensions as it is already processed.\n",
      "Skipping nmf with 25 dimensions as it is already processed.\n",
      "Skipping vanillavae with 25 dimensions as it is already processed.\n",
      "Skipping betavae with 25 dimensions as it is already processed.\n",
      "Skipping betatcvae with 25 dimensions as it is already processed.\n",
      "Skipping pca with 30 dimensions as it is already processed.\n",
      "Skipping ica with 30 dimensions as it is already processed.\n",
      "Skipping nmf with 30 dimensions as it is already processed.\n",
      "Skipping vanillavae with 30 dimensions as it is already processed.\n",
      "Skipping betavae with 30 dimensions as it is already processed.\n",
      "Skipping betatcvae with 30 dimensions as it is already processed.\n",
      "Skipping pca with 35 dimensions as it is already processed.\n",
      "Skipping ica with 35 dimensions as it is already processed.\n",
      "Skipping nmf with 35 dimensions as it is already processed.\n",
      "Skipping vanillavae with 35 dimensions as it is already processed.\n",
      "Skipping betavae with 35 dimensions as it is already processed.\n",
      "Skipping betatcvae with 35 dimensions as it is already processed.\n",
      "Skipping pca with 40 dimensions as it is already processed.\n",
      "Skipping ica with 40 dimensions as it is already processed.\n",
      "Skipping nmf with 40 dimensions as it is already processed.\n",
      "Skipping vanillavae with 40 dimensions as it is already processed.\n",
      "Skipping betavae with 40 dimensions as it is already processed.\n",
      "Skipping betatcvae with 40 dimensions as it is already processed.\n",
      "Skipping pca with 45 dimensions as it is already processed.\n",
      "Skipping ica with 45 dimensions as it is already processed.\n",
      "Skipping nmf with 45 dimensions as it is already processed.\n",
      "Skipping vanillavae with 45 dimensions as it is already processed.\n",
      "Skipping betavae with 45 dimensions as it is already processed.\n",
      "Skipping betatcvae with 45 dimensions as it is already processed.\n",
      "Skipping pca with 50 dimensions as it is already processed.\n",
      "Skipping ica with 50 dimensions as it is already processed.\n",
      "Skipping nmf with 50 dimensions as it is already processed.\n",
      "Skipping vanillavae with 50 dimensions as it is already processed.\n",
      "Skipping betavae with 50 dimensions as it is already processed.\n",
      "Skipping betatcvae with 50 dimensions as it is already processed.\n",
      "Skipping pca with 60 dimensions as it is already processed.\n",
      "Skipping ica with 60 dimensions as it is already processed.\n",
      "Skipping nmf with 60 dimensions as it is already processed.\n",
      "Skipping vanillavae with 60 dimensions as it is already processed.\n",
      "Skipping betavae with 60 dimensions as it is already processed.\n",
      "Skipping betatcvae with 60 dimensions as it is already processed.\n",
      "Skipping pca with 70 dimensions as it is already processed.\n",
      "Skipping ica with 70 dimensions as it is already processed.\n",
      "Skipping nmf with 70 dimensions as it is already processed.\n",
      "Skipping vanillavae with 70 dimensions as it is already processed.\n",
      "Skipping betavae with 70 dimensions as it is already processed.\n",
      "Skipping betatcvae with 70 dimensions as it is already processed.\n",
      "Skipping pca with 80 dimensions as it is already processed.\n",
      "Skipping ica with 80 dimensions as it is already processed.\n",
      "Skipping nmf with 80 dimensions as it is already processed.\n",
      "Skipping vanillavae with 80 dimensions as it is already processed.\n",
      "Skipping betavae with 80 dimensions as it is already processed.\n",
      "Skipping betatcvae with 80 dimensions as it is already processed.\n",
      "Skipping pca with 90 dimensions as it is already processed.\n",
      "Skipping ica with 90 dimensions as it is already processed.\n",
      "Skipping nmf with 90 dimensions as it is already processed.\n",
      "Skipping vanillavae with 90 dimensions as it is already processed.\n",
      "Skipping betavae with 90 dimensions as it is already processed.\n",
      "Skipping betatcvae with 90 dimensions as it is already processed.\n",
      "Skipping pca with 100 dimensions as it is already processed.\n",
      "Skipping ica with 100 dimensions as it is already processed.\n",
      "Skipping nmf with 100 dimensions as it is already processed.\n",
      "Skipping vanillavae with 100 dimensions as it is already processed.\n",
      "Skipping betavae with 100 dimensions as it is already processed.\n",
      "Skipping betatcvae with 100 dimensions as it is already processed.\n",
      "Skipping pca with 150 dimensions as it is already processed.\n",
      "Skipping ica with 150 dimensions as it is already processed.\n",
      "Skipping nmf with 150 dimensions as it is already processed.\n",
      "Skipping vanillavae with 150 dimensions as it is already processed.\n",
      "Skipping betavae with 150 dimensions as it is already processed.\n",
      "Skipping betatcvae with 150 dimensions as it is already processed.\n",
      "Running GSEA for pca\n",
      "          z  full_model_z model  \\\n",
      "1         0           200   pca   \n",
      "2         0           200   pca   \n",
      "3         0           200   pca   \n",
      "4         0           200   pca   \n",
      "5         0           200   pca   \n",
      "...     ...           ...   ...   \n",
      "206726  199           200   pca   \n",
      "206729  199           200   pca   \n",
      "206731  199           200   pca   \n",
      "206732  199           200   pca   \n",
      "206764  199           200   pca   \n",
      "\n",
      "                                         reactome pathway  gsea es score  \\\n",
      "1       TGF-beta Receptor Signaling Activates SMADs R-...       0.585931   \n",
      "2                  RHO GTPases Activate CIT R-HSA-5625900       0.707092   \n",
      "3       Downregulation Of TGF-beta Receptor Signaling ...       0.658069   \n",
      "4       Heparan Sulfate/Heparin (HS-GAG) Metabolism R-...      -0.711326   \n",
      "5           Interleukin-20 Family Signaling R-HSA-8854691      -0.777165   \n",
      "...                                                   ...            ...   \n",
      "206726      Nef Mediated CD8 Down-regulation R-HSA-182218       0.606340   \n",
      "206729  Suppression Of Phagosomal Maturation R-HSA-963...       0.604016   \n",
      "206731  WNT5A-dependent Internalization Of FZD2, FZD5 ...       0.603354   \n",
      "206732  WNT5A-dependent Internalization Of FZD4 R-HSA-...       0.603354   \n",
      "206764        PINK1-PRKN Mediated Mitophagy R-HSA-5205685       0.586214   \n",
      "\n",
      "        nes score   p value  shuffled  \n",
      "1        2.402254  0.016294     False  \n",
      "2        2.333252  0.019635     False  \n",
      "3        2.227189  0.025935     False  \n",
      "4       -2.209873  0.027114     False  \n",
      "5       -2.190613  0.028480     False  \n",
      "...           ...       ...       ...  \n",
      "206726   1.272259  0.203281     False  \n",
      "206729   1.261220  0.207230     False  \n",
      "206731   1.258076  0.208364     False  \n",
      "206732   1.258076  0.208364     False  \n",
      "206764   1.176716  0.239309     False  \n",
      "\n",
      "[14326 rows x 8 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/juliacurd/anaconda3/envs/gene_dependency_representations/lib/python3.12/site-packages/sklearn/decomposition/_fastica.py:128: ConvergenceWarning: FastICA did not converge. Consider increasing tolerance or the maximum number of iterations.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running GSEA for ica\n",
      "          z  full_model_z model  \\\n",
      "10        0           200   ica   \n",
      "11        0           200   ica   \n",
      "14        0           200   ica   \n",
      "15        0           200   ica   \n",
      "16        0           200   ica   \n",
      "...     ...           ...   ...   \n",
      "206756  199           200   ica   \n",
      "206763  199           200   ica   \n",
      "206770  199           200   ica   \n",
      "206781  199           200   ica   \n",
      "206783  199           200   ica   \n",
      "\n",
      "                                         reactome pathway  gsea es score  \\\n",
      "10      Resolution Of AP Sites Via Multiple-Nucleotide...      -0.587210   \n",
      "11      PCNA-Dependent Long Patch Base Excision Repair...      -0.587210   \n",
      "14      Infection With Mycobacterium Tuberculosis R-HS...       0.717561   \n",
      "15          Response Of Mtb To Phagocytosis R-HSA-9637690       0.717560   \n",
      "16      Plasma Lipoprotein Assembly, Remodeling, And C...       0.673192   \n",
      "...                                                   ...            ...   \n",
      "206756  Defective Intrinsic Pathway For Apoptosis R-HS...       0.592700   \n",
      "206763     DCC Mediated Attractive Signaling R-HSA-418885      -0.619495   \n",
      "206770  SUMOylation Of Intracellular Receptors R-HSA-4...       0.598542   \n",
      "206781  BBSome-mediated Cargo-Targeting To Cilium R-HS...       0.585822   \n",
      "206783          RHO GTPases Activate IQGAPs R-HSA-5626467      -0.601548   \n",
      "\n",
      "        nes score   p value  shuffled  \n",
      "10      -2.270205  0.023195     False  \n",
      "11      -2.270205  0.023195     False  \n",
      "14       2.239598  0.025117     False  \n",
      "15       2.239598  0.025117     False  \n",
      "16       2.228734  0.025832     False  \n",
      "...           ...       ...       ...  \n",
      "206756   1.354535  0.175566     False  \n",
      "206763  -1.333433  0.182390     False  \n",
      "206770   1.312145  0.189471     False  \n",
      "206781   1.250053  0.211280     False  \n",
      "206783  -1.247760  0.212119     False  \n",
      "\n",
      "[15224 rows x 8 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/juliacurd/anaconda3/envs/gene_dependency_representations/lib/python3.12/site-packages/sklearn/decomposition/_nmf.py:1759: ConvergenceWarning: Maximum number of iterations 1000 reached. Increase it to improve convergence.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running GSEA for nmf\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[I 2024-09-20 11:24:46,848] A new study created in memory with name: no-name-7a995e52-991e-42d8-b057-e25c79ca6427\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          z  full_model_z model  \\\n",
      "0         0           200   nmf   \n",
      "1         0           200   nmf   \n",
      "2         0           200   nmf   \n",
      "4         0           200   nmf   \n",
      "5         0           200   nmf   \n",
      "...     ...           ...   ...   \n",
      "206645  199           200   nmf   \n",
      "206654  199           200   nmf   \n",
      "206660  199           200   nmf   \n",
      "206690  199           200   nmf   \n",
      "206715  199           200   nmf   \n",
      "\n",
      "                                         reactome pathway  gsea es score  \\\n",
      "0                 Signaling By CSF3 (G-CSF) R-HSA-9674555       0.682296   \n",
      "1                      Signaling By SCF-KIT R-HSA-1433557       0.705609   \n",
      "2       Inactivation Of CSF3 (G-CSF) Signaling R-HSA-9...       0.709191   \n",
      "4                 FLT3 Signaling In Disease R-HSA-9682385       0.694062   \n",
      "5             Interleukin-2 Family Signaling R-HSA-451927       0.690910   \n",
      "...                                                   ...            ...   \n",
      "206645          RHO GTPases Activate IQGAPs R-HSA-5626467       0.663104   \n",
      "206654              Elastic Fibre Formation R-HSA-1566948       0.654520   \n",
      "206660  Synthesis Of PIPs At Golgi Membrane R-HSA-1660514       0.615921   \n",
      "206690  Uptake And Function Of Anthrax Toxins R-HSA-52...       0.587482   \n",
      "206715       Adherens Junctions Interactions R-HSA-418990       0.605522   \n",
      "\n",
      "        nes score   p value  shuffled  \n",
      "0        3.014830  0.002571     False  \n",
      "1        2.819561  0.004809     False  \n",
      "2        2.739733  0.006149     False  \n",
      "4        2.654265  0.007948     False  \n",
      "5        2.636338  0.008381     False  \n",
      "...           ...       ...       ...  \n",
      "206645   1.648483  0.099253     False  \n",
      "206654   1.606073  0.108258     False  \n",
      "206660   1.569448  0.116543     False  \n",
      "206690   1.419485  0.155758     False  \n",
      "206715   1.361546  0.173341     False  \n",
      "\n",
      "[10770 rows x 8 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[I 2024-09-20 11:25:19,301] Trial 0 finished with value: 46.82982305270522 and parameters: {'learning_rate': 0.0017584585349813963, 'batch_size': 78, 'epochs': 776, 'optimizer_type': 'rmsprop'}. Best is trial 0 with value: 46.82982305270522.\n",
      "[I 2024-09-20 11:25:47,335] Trial 1 finished with value: 46.72542888584422 and parameters: {'learning_rate': 0.004255308785424037, 'batch_size': 97, 'epochs': 621, 'optimizer_type': 'adam'}. Best is trial 1 with value: 46.72542888584422.\n",
      "[I 2024-09-20 11:26:48,484] Trial 2 finished with value: 46.77761859039762 and parameters: {'learning_rate': 0.0011923529493946341, 'batch_size': 33, 'epochs': 806, 'optimizer_type': 'rmsprop'}. Best is trial 1 with value: 46.72542888584422.\n",
      "[I 2024-09-20 11:27:14,930] Trial 3 finished with value: 46.86338920023904 and parameters: {'learning_rate': 0.0023892918163823003, 'batch_size': 94, 'epochs': 643, 'optimizer_type': 'rmsprop'}. Best is trial 1 with value: 46.72542888584422.\n",
      "[I 2024-09-20 11:27:30,459] Trial 4 finished with value: 47.072908771571825 and parameters: {'learning_rate': 0.0017569503620775556, 'batch_size': 87, 'epochs': 318, 'optimizer_type': 'adam'}. Best is trial 1 with value: 46.72542888584422.\n",
      "[I 2024-09-20 11:27:42,226] Trial 5 finished with value: 166.94499256646455 and parameters: {'learning_rate': 3.5889230378967945e-05, 'batch_size': 71, 'epochs': 227, 'optimizer_type': 'adam'}. Best is trial 1 with value: 46.72542888584422.\n",
      "[I 2024-09-20 11:28:22,122] Trial 6 finished with value: 59.80279923624067 and parameters: {'learning_rate': 0.00013329206513994988, 'batch_size': 103, 'epochs': 899, 'optimizer_type': 'adam'}. Best is trial 1 with value: 46.72542888584422.\n",
      "[I 2024-09-20 11:28:24,646] Trial 7 finished with value: 47.35780165942747 and parameters: {'learning_rate': 0.003747375013080491, 'batch_size': 17, 'epochs': 21, 'optimizer_type': 'adam'}. Best is trial 1 with value: 46.72542888584422.\n",
      "[I 2024-09-20 11:29:32,903] Trial 8 finished with value: 46.78051520959655 and parameters: {'learning_rate': 0.0020346946992894246, 'batch_size': 24, 'epochs': 746, 'optimizer_type': 'adam'}. Best is trial 1 with value: 46.72542888584422.\n",
      "[I 2024-09-20 11:29:46,671] Trial 9 finished with value: 47.01088921845849 and parameters: {'learning_rate': 0.004362468024398511, 'batch_size': 74, 'epochs': 275, 'optimizer_type': 'adam'}. Best is trial 1 with value: 46.72542888584422.\n",
      "[I 2024-09-20 11:30:13,647] Trial 10 finished with value: 46.90885688440124 and parameters: {'learning_rate': 0.003327920852494344, 'batch_size': 51, 'epochs': 561, 'optimizer_type': 'rmsprop'}. Best is trial 1 with value: 46.72542888584422.\n",
      "[I 2024-09-20 11:31:22,645] Trial 11 finished with value: 46.88896274851329 and parameters: {'learning_rate': 0.004937627292988565, 'batch_size': 47, 'epochs': 1000, 'optimizer_type': 'rmsprop'}. Best is trial 1 with value: 46.72542888584422.\n",
      "[I 2024-09-20 11:31:37,578] Trial 12 finished with value: 47.00296612639925 and parameters: {'learning_rate': 0.000948954235916621, 'batch_size': 112, 'epochs': 440, 'optimizer_type': 'rmsprop'}. Best is trial 1 with value: 46.72542888584422.\n",
      "[I 2024-09-20 11:32:34,764] Trial 13 finished with value: 46.71393715018657 and parameters: {'learning_rate': 0.003173785577502527, 'batch_size': 38, 'epochs': 760, 'optimizer_type': 'rmsprop'}. Best is trial 13 with value: 46.71393715018657.\n",
      "[I 2024-09-20 11:33:12,543] Trial 14 finished with value: 46.960374701201026 and parameters: {'learning_rate': 0.0031849387855634294, 'batch_size': 52, 'epochs': 641, 'optimizer_type': 'adam'}. Best is trial 13 with value: 46.71393715018657.\n",
      "[I 2024-09-20 11:33:40,883] Trial 15 finished with value: 46.77599910906891 and parameters: {'learning_rate': 0.0039133020672555766, 'batch_size': 37, 'epochs': 454, 'optimizer_type': 'rmsprop'}. Best is trial 13 with value: 46.71393715018657.\n",
      "[I 2024-09-20 11:34:15,203] Trial 16 finished with value: 46.79382433535448 and parameters: {'learning_rate': 0.0028978005252612567, 'batch_size': 62, 'epochs': 635, 'optimizer_type': 'adam'}. Best is trial 13 with value: 46.71393715018657.\n",
      "[I 2024-09-20 11:35:01,465] Trial 17 finished with value: 46.7957788268132 and parameters: {'learning_rate': 0.004938069112711586, 'batch_size': 59, 'epochs': 890, 'optimizer_type': 'rmsprop'}. Best is trial 13 with value: 46.71393715018657.\n",
      "[I 2024-09-20 11:35:35,657] Trial 18 finished with value: 46.850909196085006 and parameters: {'learning_rate': 0.00413988814215623, 'batch_size': 39, 'epochs': 551, 'optimizer_type': 'rmsprop'}. Best is trial 13 with value: 46.71393715018657.\n",
      "[I 2024-09-20 11:35:53,790] Trial 19 finished with value: 47.98253301364272 and parameters: {'learning_rate': 0.0026856315987842955, 'batch_size': 85, 'epochs': 392, 'optimizer_type': 'adam'}. Best is trial 13 with value: 46.71393715018657.\n",
      "[I 2024-09-20 11:36:17,427] Trial 20 finished with value: 47.23522621268657 and parameters: {'learning_rate': 0.004522159713115483, 'batch_size': 112, 'epochs': 703, 'optimizer_type': 'rmsprop'}. Best is trial 13 with value: 46.71393715018657.\n",
      "[I 2024-09-20 11:36:50,913] Trial 21 finished with value: 46.91123156191698 and parameters: {'learning_rate': 0.0036821755685764656, 'batch_size': 33, 'epochs': 494, 'optimizer_type': 'rmsprop'}. Best is trial 13 with value: 46.71393715018657.\n",
      "[I 2024-09-20 11:37:17,108] Trial 22 finished with value: 46.92678587045243 and parameters: {'learning_rate': 0.00380332401254952, 'batch_size': 40, 'epochs': 499, 'optimizer_type': 'rmsprop'}. Best is trial 13 with value: 46.71393715018657.\n",
      "[I 2024-09-20 11:37:25,604] Trial 23 finished with value: 47.04247618717934 and parameters: {'learning_rate': 0.0033475040206617326, 'batch_size': 25, 'epochs': 112, 'optimizer_type': 'rmsprop'}. Best is trial 13 with value: 46.71393715018657.\n",
      "[I 2024-09-20 11:37:44,011] Trial 24 finished with value: 46.90310113252099 and parameters: {'learning_rate': 0.004108559384862344, 'batch_size': 46, 'epochs': 370, 'optimizer_type': 'rmsprop'}. Best is trial 13 with value: 46.71393715018657.\n",
      "[I 2024-09-20 11:38:18,616] Trial 25 finished with value: 46.854180999300375 and parameters: {'learning_rate': 0.004555481264043422, 'batch_size': 67, 'epochs': 866, 'optimizer_type': 'rmsprop'}. Best is trial 13 with value: 46.71393715018657.\n",
      "[I 2024-09-20 11:38:50,143] Trial 26 finished with value: 46.85686080704874 and parameters: {'learning_rate': 0.0031061970585689446, 'batch_size': 55, 'epochs': 567, 'optimizer_type': 'adam'}. Best is trial 13 with value: 46.71393715018657.\n",
      "[I 2024-09-20 11:39:57,725] Trial 27 finished with value: 46.791016262680735 and parameters: {'learning_rate': 0.00359180280251854, 'batch_size': 30, 'epochs': 699, 'optimizer_type': 'rmsprop'}. Best is trial 13 with value: 46.71393715018657.\n",
      "[I 2024-09-20 11:41:00,207] Trial 28 finished with value: 46.81276254226912 and parameters: {'learning_rate': 0.003961990875907444, 'batch_size': 40, 'epochs': 987, 'optimizer_type': 'adam'}. Best is trial 13 with value: 46.71393715018657.\n",
      "[I 2024-09-20 11:43:44,029] Trial 29 finished with value: 46.80767193694613 and parameters: {'learning_rate': 0.0024052865553956288, 'batch_size': 16, 'epochs': 798, 'optimizer_type': 'rmsprop'}. Best is trial 13 with value: 46.71393715018657.\n",
      "[I 2024-09-20 11:44:01,349] Trial 30 finished with value: 46.67599751486707 and parameters: {'learning_rate': 0.0028640627440416045, 'batch_size': 81, 'epochs': 447, 'optimizer_type': 'rmsprop'}. Best is trial 30 with value: 46.67599751486707.\n",
      "[I 2024-09-20 11:44:19,429] Trial 31 finished with value: 46.88536704903218 and parameters: {'learning_rate': 0.002806443248510283, 'batch_size': 82, 'epochs': 436, 'optimizer_type': 'rmsprop'}. Best is trial 30 with value: 46.67599751486707.\n",
      "[I 2024-09-20 11:44:41,873] Trial 32 finished with value: 46.89202452702309 and parameters: {'learning_rate': 0.0034634869027957097, 'batch_size': 93, 'epochs': 579, 'optimizer_type': 'rmsprop'}. Best is trial 30 with value: 46.67599751486707.\n",
      "[I 2024-09-20 11:44:58,919] Trial 33 finished with value: 46.81755917677239 and parameters: {'learning_rate': 0.0030063327205201063, 'batch_size': 96, 'epochs': 460, 'optimizer_type': 'rmsprop'}. Best is trial 30 with value: 46.67599751486707.\n",
      "[I 2024-09-20 11:45:07,084] Trial 34 finished with value: 47.43842390829058 and parameters: {'learning_rate': 0.0016183362754050054, 'batch_size': 78, 'epochs': 199, 'optimizer_type': 'rmsprop'}. Best is trial 30 with value: 46.67599751486707.\n",
      "[I 2024-09-20 11:45:33,672] Trial 35 finished with value: 47.0321923099347 and parameters: {'learning_rate': 0.0021439069541941236, 'batch_size': 103, 'epochs': 714, 'optimizer_type': 'rmsprop'}. Best is trial 30 with value: 46.67599751486707.\n",
      "[I 2024-09-20 11:45:47,239] Trial 36 finished with value: 47.0585742551889 and parameters: {'learning_rate': 0.004135139683604476, 'batch_size': 90, 'epochs': 345, 'optimizer_type': 'rmsprop'}. Best is trial 30 with value: 46.67599751486707.\n",
      "[I 2024-09-20 11:46:15,802] Trial 37 finished with value: 46.85812405258862 and parameters: {'learning_rate': 0.0026805977295566285, 'batch_size': 100, 'epochs': 626, 'optimizer_type': 'adam'}. Best is trial 30 with value: 46.67599751486707.\n",
      "[I 2024-09-20 11:46:29,404] Trial 38 finished with value: 46.95040137447528 and parameters: {'learning_rate': 0.004662801891250655, 'batch_size': 80, 'epochs': 275, 'optimizer_type': 'adam'}. Best is trial 30 with value: 46.67599751486707.\n",
      "[I 2024-09-20 11:47:04,627] Trial 39 finished with value: 46.94814598880597 and parameters: {'learning_rate': 0.0042999422438919094, 'batch_size': 69, 'epochs': 823, 'optimizer_type': 'rmsprop'}. Best is trial 30 with value: 46.67599751486707.\n",
      "[I 2024-09-20 11:47:46,601] Trial 40 finished with value: 46.85988790028131 and parameters: {'learning_rate': 0.003889943301352433, 'batch_size': 23, 'epochs': 418, 'optimizer_type': 'adam'}. Best is trial 30 with value: 46.67599751486707.\n",
      "[I 2024-09-20 11:48:45,928] Trial 41 finished with value: 46.90701972619811 and parameters: {'learning_rate': 0.0014024180905982337, 'batch_size': 33, 'epochs': 779, 'optimizer_type': 'rmsprop'}. Best is trial 30 with value: 46.67599751486707.\n",
      "[I 2024-09-20 11:49:17,604] Trial 42 finished with value: 46.920481303200795 and parameters: {'learning_rate': 0.0008221099096857739, 'batch_size': 73, 'epochs': 735, 'optimizer_type': 'rmsprop'}. Best is trial 30 with value: 46.67599751486707.\n",
      "[I 2024-09-20 11:50:07,539] Trial 43 finished with value: 46.74882138380364 and parameters: {'learning_rate': 0.0007858092463583612, 'batch_size': 46, 'epochs': 924, 'optimizer_type': 'rmsprop'}. Best is trial 30 with value: 46.67599751486707.\n",
      "[I 2024-09-20 11:51:06,834] Trial 44 finished with value: 46.937883883803636 and parameters: {'learning_rate': 0.002231486913335994, 'batch_size': 46, 'epochs': 951, 'optimizer_type': 'rmsprop'}. Best is trial 30 with value: 46.67599751486707.\n",
      "[I 2024-09-20 11:51:58,929] Trial 45 finished with value: 46.8697778502507 and parameters: {'learning_rate': 0.0005413273844206896, 'batch_size': 36, 'epochs': 843, 'optimizer_type': 'rmsprop'}. Best is trial 30 with value: 46.67599751486707.\n",
      "[I 2024-09-20 11:52:23,239] Trial 46 finished with value: 46.956228683243936 and parameters: {'learning_rate': 0.001751198471686076, 'batch_size': 57, 'epochs': 517, 'optimizer_type': 'rmsprop'}. Best is trial 30 with value: 46.67599751486707.\n",
      "[I 2024-09-20 11:53:07,140] Trial 47 finished with value: 46.87226339311742 and parameters: {'learning_rate': 0.002552804672766833, 'batch_size': 44, 'epochs': 668, 'optimizer_type': 'adam'}. Best is trial 30 with value: 46.67599751486707.\n",
      "[I 2024-09-20 11:54:03,946] Trial 48 finished with value: 46.95863337730294 and parameters: {'learning_rate': 0.0032440955713366982, 'batch_size': 50, 'epochs': 946, 'optimizer_type': 'rmsprop'}. Best is trial 30 with value: 46.67599751486707.\n",
      "[I 2024-09-20 11:54:35,599] Trial 49 finished with value: 46.81807952425373 and parameters: {'learning_rate': 0.0034980547703164755, 'batch_size': 64, 'epochs': 595, 'optimizer_type': 'adam'}. Best is trial 30 with value: 46.67599751486707.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running GSEA for vanillavae\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[I 2024-09-20 11:58:53,962] A new study created in memory with name: no-name-cbd38c39-4890-4b10-a787-6563a7aa4b14\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          z  full_model_z       model  \\\n",
      "0         1           200  vanillavae   \n",
      "1         1           200  vanillavae   \n",
      "3         1           200  vanillavae   \n",
      "13        1           200  vanillavae   \n",
      "14        1           200  vanillavae   \n",
      "...     ...           ...         ...   \n",
      "206733  200           200  vanillavae   \n",
      "206738  200           200  vanillavae   \n",
      "206739  200           200  vanillavae   \n",
      "206744  200           200  vanillavae   \n",
      "206759  200           200  vanillavae   \n",
      "\n",
      "                                         reactome pathway  gsea es score  \\\n",
      "0       Downregulation Of TGF-beta Receptor Signaling ...      -0.731404   \n",
      "1       Translesion Synthesis By Y Family DNA Polymera...      -0.600198   \n",
      "3          Eukaryotic Translation Termination R-HSA-72764      -0.880339   \n",
      "13                               Methylation R-HSA-156581      -0.743081   \n",
      "14      Estrogen-dependent Nuclear Events Downstream O...      -0.666458   \n",
      "...                                                   ...            ...   \n",
      "206733          CASP8 Activity Is Inhibited R-HSA-5218900       0.613026   \n",
      "206738   Translation Of Structural Proteins R-HSA-9683701      -0.603529   \n",
      "206739  NF-kB Is Activated And Signals Survival R-HSA-...       0.594558   \n",
      "206744  Regulation Of Gene Expression In Beta Cells R-...       0.607863   \n",
      "206759            PTK6 Regulates Cell Cycle R-HSA-8849470       0.601326   \n",
      "\n",
      "        nes score   p value  shuffled  \n",
      "0       -2.674463  0.007485     False  \n",
      "1       -2.618370  0.008835     False  \n",
      "3       -2.427116  0.015219     False  \n",
      "13      -2.167898  0.030166     False  \n",
      "14      -2.135118  0.032751     False  \n",
      "...           ...       ...       ...  \n",
      "206733   1.243768  0.213585     False  \n",
      "206738  -1.235652  0.216588     False  \n",
      "206739   1.235268  0.216731     False  \n",
      "206744   1.219581  0.222624     False  \n",
      "206759   1.188964  0.234454     False  \n",
      "\n",
      "[12395 rows x 8 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[I 2024-09-20 11:59:12,373] Trial 0 finished with value: 47.58187711345616 and parameters: {'beta': 7.493976013844335, 'learning_rate': 0.003898074223506265, 'batch_size': 62, 'epochs': 417, 'optimizer_type': 'rmsprop'}. Best is trial 0 with value: 47.58187711345616.\n",
      "[I 2024-09-20 12:00:03,609] Trial 1 finished with value: 47.564102400594685 and parameters: {'beta': 6.724034283118957, 'learning_rate': 0.00047403081387432293, 'batch_size': 46, 'epochs': 958, 'optimizer_type': 'rmsprop'}. Best is trial 1 with value: 47.564102400594685.\n",
      "[I 2024-09-20 12:00:17,827] Trial 2 finished with value: 47.678801488164645 and parameters: {'beta': 5.943240146082806, 'learning_rate': 0.00413096225697889, 'batch_size': 40, 'epochs': 208, 'optimizer_type': 'adam'}. Best is trial 1 with value: 47.564102400594685.\n",
      "[I 2024-09-20 12:00:40,027] Trial 3 finished with value: 47.58383574983967 and parameters: {'beta': 3.528588949400702, 'learning_rate': 0.004513367866099909, 'batch_size': 95, 'epochs': 583, 'optimizer_type': 'rmsprop'}. Best is trial 1 with value: 47.564102400594685.\n",
      "[I 2024-09-20 12:01:02,000] Trial 4 finished with value: 47.57088021805037 and parameters: {'beta': 8.877744545271772, 'learning_rate': 0.001117833466340458, 'batch_size': 76, 'epochs': 559, 'optimizer_type': 'rmsprop'}. Best is trial 1 with value: 47.564102400594685.\n",
      "[I 2024-09-20 12:01:19,616] Trial 5 finished with value: 48.09697666452892 and parameters: {'beta': 8.543335344727758, 'learning_rate': 0.0014280680238737894, 'batch_size': 94, 'epochs': 384, 'optimizer_type': 'adam'}. Best is trial 1 with value: 47.564102400594685.\n",
      "[I 2024-09-20 12:02:22,856] Trial 6 finished with value: 47.57641460361766 and parameters: {'beta': 4.714121033431499, 'learning_rate': 0.0021996463001759466, 'batch_size': 39, 'epochs': 899, 'optimizer_type': 'adam'}. Best is trial 1 with value: 47.564102400594685.\n",
      "[I 2024-09-20 12:02:39,429] Trial 7 finished with value: 48.20941289645523 and parameters: {'beta': 9.14020768628416, 'learning_rate': 0.0035353585728753606, 'batch_size': 49, 'epochs': 280, 'optimizer_type': 'adam'}. Best is trial 1 with value: 47.564102400594685.\n",
      "[I 2024-09-20 12:02:58,639] Trial 8 finished with value: 47.57843655258862 and parameters: {'beta': 8.592988928433284, 'learning_rate': 0.0017135617655002974, 'batch_size': 59, 'epochs': 354, 'optimizer_type': 'adam'}. Best is trial 1 with value: 47.564102400594685.\n",
      "[I 2024-09-20 12:03:27,825] Trial 9 finished with value: 47.32207723589086 and parameters: {'beta': 1.5006046317276556, 'learning_rate': 0.0027856990960770877, 'batch_size': 87, 'epochs': 779, 'optimizer_type': 'rmsprop'}. Best is trial 9 with value: 47.32207723589086.\n",
      "[I 2024-09-20 12:06:07,451] Trial 10 finished with value: 47.5936303893132 and parameters: {'beta': 1.521528282617461, 'learning_rate': 0.0029129343276470972, 'batch_size': 16, 'epochs': 753, 'optimizer_type': 'rmsprop'}. Best is trial 9 with value: 47.32207723589086.\n",
      "[I 2024-09-20 12:06:42,580] Trial 11 finished with value: 47.59244812922691 and parameters: {'beta': 2.366076663045707, 'learning_rate': 0.00034811189022527275, 'batch_size': 109, 'epochs': 970, 'optimizer_type': 'rmsprop'}. Best is trial 9 with value: 47.32207723589086.\n",
      "[I 2024-09-20 12:07:14,389] Trial 12 finished with value: 136.93602622871967 and parameters: {'beta': 6.453490682577276, 'learning_rate': 3.6191019445420785e-05, 'batch_size': 82, 'epochs': 812, 'optimizer_type': 'rmsprop'}. Best is trial 9 with value: 47.32207723589086.\n",
      "[I 2024-09-20 12:07:15,425] Trial 13 finished with value: 92.04499712132696 and parameters: {'beta': 4.391814934732197, 'learning_rate': 0.0028257038305040632, 'batch_size': 73, 'epochs': 24, 'optimizer_type': 'rmsprop'}. Best is trial 9 with value: 47.32207723589086.\n",
      "[I 2024-09-20 12:08:34,850] Trial 14 finished with value: 47.608928987873135 and parameters: {'beta': 7.224212128943571, 'learning_rate': 0.004991547174631975, 'batch_size': 27, 'epochs': 697, 'optimizer_type': 'rmsprop'}. Best is trial 9 with value: 47.32207723589086.\n",
      "[I 2024-09-20 12:09:18,381] Trial 15 finished with value: 47.568514969099816 and parameters: {'beta': 3.574284979723586, 'learning_rate': 0.0006372277362869231, 'batch_size': 50, 'epochs': 873, 'optimizer_type': 'rmsprop'}. Best is trial 9 with value: 47.32207723589086.\n",
      "[I 2024-09-20 12:09:45,179] Trial 16 finished with value: 47.3781543333139 and parameters: {'beta': 1.7484600969023072, 'learning_rate': 0.002167950913133399, 'batch_size': 92, 'epochs': 700, 'optimizer_type': 'rmsprop'}. Best is trial 9 with value: 47.32207723589086.\n",
      "[I 2024-09-20 12:10:08,030] Trial 17 finished with value: 47.067790199393656 and parameters: {'beta': 1.2891686298000193, 'learning_rate': 0.002174393477726852, 'batch_size': 112, 'epochs': 656, 'optimizer_type': 'rmsprop'}. Best is trial 17 with value: 47.067790199393656.\n",
      "[I 2024-09-20 12:10:29,443] Trial 18 finished with value: 47.02964595965485 and parameters: {'beta': 1.0796848971719937, 'learning_rate': 0.0030815320568341655, 'batch_size': 112, 'epochs': 619, 'optimizer_type': 'rmsprop'}. Best is trial 18 with value: 47.02964595965485.\n",
      "[I 2024-09-20 12:10:50,423] Trial 19 finished with value: 47.57560415694963 and parameters: {'beta': 2.749210095491491, 'learning_rate': 0.0031865826494410065, 'batch_size': 112, 'epochs': 620, 'optimizer_type': 'rmsprop'}. Best is trial 18 with value: 47.02964595965485.\n",
      "[I 2024-09-20 12:11:07,048] Trial 20 finished with value: 47.56567765420942 and parameters: {'beta': 2.724562114672014, 'learning_rate': 0.0022748482717566848, 'batch_size': 105, 'epochs': 465, 'optimizer_type': 'rmsprop'}. Best is trial 18 with value: 47.02964595965485.\n",
      "[I 2024-09-20 12:11:31,826] Trial 21 finished with value: 47.06839508512127 and parameters: {'beta': 1.1310608053468103, 'learning_rate': 0.0032873316404261584, 'batch_size': 100, 'epochs': 702, 'optimizer_type': 'rmsprop'}. Best is trial 18 with value: 47.02964595965485.\n",
      "[I 2024-09-20 12:11:55,283] Trial 22 finished with value: 47.13910003206623 and parameters: {'beta': 1.2439757441265746, 'learning_rate': 0.003509096899137333, 'batch_size': 102, 'epochs': 650, 'optimizer_type': 'rmsprop'}. Best is trial 18 with value: 47.02964595965485.\n",
      "[I 2024-09-20 12:12:14,144] Trial 23 finished with value: 46.97450771047108 and parameters: {'beta': 1.159058447114301, 'learning_rate': 0.0018122749492689998, 'batch_size': 101, 'epochs': 533, 'optimizer_type': 'rmsprop'}. Best is trial 23 with value: 46.97450771047108.\n",
      "[I 2024-09-20 12:12:32,334] Trial 24 finished with value: 47.57008876515858 and parameters: {'beta': 2.2937741211385125, 'learning_rate': 0.0018765989455366834, 'batch_size': 112, 'epochs': 532, 'optimizer_type': 'rmsprop'}. Best is trial 23 with value: 46.97450771047108.\n",
      "[I 2024-09-20 12:12:49,676] Trial 25 finished with value: 47.568375772504666 and parameters: {'beta': 3.2112245514104245, 'learning_rate': 0.0012968572545944167, 'batch_size': 98, 'epochs': 488, 'optimizer_type': 'rmsprop'}. Best is trial 23 with value: 46.97450771047108.\n",
      "[I 2024-09-20 12:13:17,805] Trial 26 finished with value: 53.030890712453356 and parameters: {'beta': 4.349093125665837, 'learning_rate': 0.002492881164281678, 'batch_size': 86, 'epochs': 628, 'optimizer_type': 'adam'}. Best is trial 23 with value: 46.97450771047108.\n",
      "[I 2024-09-20 12:13:28,633] Trial 27 finished with value: 47.65505917677239 and parameters: {'beta': 2.315798224731494, 'learning_rate': 0.0018534128514948026, 'batch_size': 105, 'epochs': 301, 'optimizer_type': 'rmsprop'}. Best is trial 23 with value: 46.97450771047108.\n",
      "[I 2024-09-20 12:13:47,309] Trial 28 finished with value: 46.93543965353894 and parameters: {'beta': 1.011291029376867, 'learning_rate': 0.000901741268442609, 'batch_size': 72, 'epochs': 454, 'optimizer_type': 'rmsprop'}. Best is trial 28 with value: 46.93543965353894.\n",
      "[I 2024-09-20 12:14:05,569] Trial 29 finished with value: 47.565653057952424 and parameters: {'beta': 9.932951355526807, 'learning_rate': 0.0011095317798884233, 'batch_size': 71, 'epochs': 443, 'optimizer_type': 'rmsprop'}. Best is trial 28 with value: 46.93543965353894.\n",
      "[I 2024-09-20 12:14:12,269] Trial 30 finished with value: 48.2169482786264 and parameters: {'beta': 1.9815990822302285, 'learning_rate': 0.0008711031999553123, 'batch_size': 64, 'epochs': 153, 'optimizer_type': 'rmsprop'}. Best is trial 28 with value: 46.93543965353894.\n",
      "[I 2024-09-20 12:14:32,065] Trial 31 finished with value: 46.81219919688666 and parameters: {'beta': 1.0242554777033308, 'learning_rate': 0.0016357088390257849, 'batch_size': 105, 'epochs': 554, 'optimizer_type': 'rmsprop'}. Best is trial 31 with value: 46.81219919688666.\n",
      "[I 2024-09-20 12:14:51,768] Trial 32 finished with value: 46.94178393919076 and parameters: {'beta': 1.002751028113356, 'learning_rate': 0.00158692483828857, 'batch_size': 89, 'epochs': 517, 'optimizer_type': 'rmsprop'}. Best is trial 31 with value: 46.81219919688666.\n",
      "[I 2024-09-20 12:15:10,798] Trial 33 finished with value: 46.89927340834888 and parameters: {'beta': 1.006934646601661, 'learning_rate': 0.0014519143944222463, 'batch_size': 88, 'epochs': 512, 'optimizer_type': 'rmsprop'}. Best is trial 31 with value: 46.81219919688666.\n",
      "[I 2024-09-20 12:15:26,617] Trial 34 finished with value: 47.63096832730877 and parameters: {'beta': 1.999962686495178, 'learning_rate': 0.00155684019772738, 'batch_size': 80, 'epochs': 401, 'optimizer_type': 'rmsprop'}. Best is trial 31 with value: 46.81219919688666.\n",
      "[I 2024-09-20 12:15:46,739] Trial 35 finished with value: 47.80712453358209 and parameters: {'beta': 2.959432403636689, 'learning_rate': 0.0006736762251428929, 'batch_size': 67, 'epochs': 489, 'optimizer_type': 'rmsprop'}. Best is trial 31 with value: 46.81219919688666.\n",
      "[I 2024-09-20 12:16:00,748] Trial 36 finished with value: 47.75779847102379 and parameters: {'beta': 3.6300259129022305, 'learning_rate': 0.0009910637852328678, 'batch_size': 90, 'epochs': 315, 'optimizer_type': 'adam'}. Best is trial 31 with value: 46.81219919688666.\n",
      "[I 2024-09-20 12:16:23,074] Trial 37 finished with value: 47.56987104273554 and parameters: {'beta': 5.48832609194002, 'learning_rate': 0.0013589527230879295, 'batch_size': 80, 'epochs': 570, 'optimizer_type': 'rmsprop'}. Best is trial 31 with value: 46.81219919688666.\n",
      "[I 2024-09-20 12:16:47,409] Trial 38 finished with value: 47.77911108216243 and parameters: {'beta': 1.891613532779381, 'learning_rate': 0.0003931812258521207, 'batch_size': 55, 'epochs': 436, 'optimizer_type': 'adam'}. Best is trial 31 with value: 46.81219919688666.\n",
      "[I 2024-09-20 12:16:56,661] Trial 39 finished with value: 47.000226285564366 and parameters: {'beta': 1.028178903975183, 'learning_rate': 0.0015188089054041282, 'batch_size': 76, 'epochs': 237, 'optimizer_type': 'rmsprop'}. Best is trial 31 with value: 46.81219919688666.\n",
      "[I 2024-09-20 12:17:09,297] Trial 40 finished with value: 47.59112202658582 and parameters: {'beta': 5.230690659885815, 'learning_rate': 0.0011905272943748537, 'batch_size': 96, 'epochs': 357, 'optimizer_type': 'rmsprop'}. Best is trial 31 with value: 46.81219919688666.\n",
      "[I 2024-09-20 12:17:28,897] Trial 41 finished with value: 47.40292385727612 and parameters: {'beta': 1.6703816868589145, 'learning_rate': 0.0017897172215139683, 'batch_size': 87, 'epochs': 517, 'optimizer_type': 'rmsprop'}. Best is trial 31 with value: 46.81219919688666.\n",
      "[I 2024-09-20 12:17:49,028] Trial 42 finished with value: 46.773943454116136 and parameters: {'beta': 1.0163030459223241, 'learning_rate': 0.001993610739621247, 'batch_size': 92, 'epochs': 540, 'optimizer_type': 'rmsprop'}. Best is trial 42 with value: 46.773943454116136.\n",
      "[I 2024-09-20 12:18:10,311] Trial 43 finished with value: 47.57383942105877 and parameters: {'beta': 2.329236963924867, 'learning_rate': 0.002537434733718447, 'batch_size': 91, 'epochs': 575, 'optimizer_type': 'rmsprop'}. Best is trial 42 with value: 46.773943454116136.\n",
      "[I 2024-09-20 12:18:25,491] Trial 44 finished with value: 47.47870228326143 and parameters: {'beta': 1.6936394997623656, 'learning_rate': 0.0008287343369103378, 'batch_size': 82, 'epochs': 386, 'optimizer_type': 'rmsprop'}. Best is trial 42 with value: 46.773943454116136.\n",
      "[I 2024-09-20 12:18:48,195] Trial 45 finished with value: 47.599915097364736 and parameters: {'beta': 1.4787894223070788, 'learning_rate': 0.002047926705753281, 'batch_size': 69, 'epochs': 468, 'optimizer_type': 'adam'}. Best is trial 42 with value: 46.773943454116136.\n",
      "[I 2024-09-20 12:19:06,560] Trial 46 finished with value: 47.58508191464552 and parameters: {'beta': 7.934337156950912, 'learning_rate': 0.001545965502622667, 'batch_size': 96, 'epochs': 521, 'optimizer_type': 'rmsprop'}. Best is trial 42 with value: 46.773943454116136.\n",
      "[I 2024-09-20 12:19:29,386] Trial 47 finished with value: 47.57150186567164 and parameters: {'beta': 2.146176745272485, 'learning_rate': 0.0024504446657392814, 'batch_size': 75, 'epochs': 586, 'optimizer_type': 'rmsprop'}. Best is trial 42 with value: 46.773943454116136.\n",
      "[I 2024-09-20 12:19:45,533] Trial 48 finished with value: 83.4684912838153 and parameters: {'beta': 2.552828308463386, 'learning_rate': 0.00013439449118390597, 'batch_size': 85, 'epochs': 429, 'optimizer_type': 'rmsprop'}. Best is trial 42 with value: 46.773943454116136.\n",
      "[I 2024-09-20 12:20:00,608] Trial 49 finished with value: 47.95057026877332 and parameters: {'beta': 1.5653796030472915, 'learning_rate': 0.0016290770362855389, 'batch_size': 105, 'epochs': 348, 'optimizer_type': 'adam'}. Best is trial 42 with value: 46.773943454116136.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running GSEA for betavae\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[I 2024-09-20 12:24:20,234] A new study created in memory with name: no-name-0bf9fd92-6d45-4375-a289-ff53a8e2b885\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          z  full_model_z    model  \\\n",
      "0         1           200  betavae   \n",
      "2         1           200  betavae   \n",
      "4         1           200  betavae   \n",
      "8         1           200  betavae   \n",
      "16        1           200  betavae   \n",
      "...     ...           ...      ...   \n",
      "206683  200           200  betavae   \n",
      "206690  200           200  betavae   \n",
      "206696  200           200  betavae   \n",
      "206707  200           200  betavae   \n",
      "206753  200           200  betavae   \n",
      "\n",
      "                                         reactome pathway  gsea es score  \\\n",
      "0                  Visual Phototransduction R-HSA-2187338      -0.753047   \n",
      "2       VEGFR2 Mediated Vascular Permeability R-HSA-52...      -0.659029   \n",
      "4       Regulation Of TP53 Activity Thru Acetylation R...      -0.612920   \n",
      "8                   Cholesterol Biosynthesis R-HSA-191273       0.660623   \n",
      "16        Downregulation Of ERBB2 Signaling R-HSA-8863795       0.615898   \n",
      "...                                                   ...            ...   \n",
      "206683  IRAK2 Mediated Activation Of TAK1 Complex Upon...       0.628215   \n",
      "206690  Role Of Phospholipids In Phagocytosis R-HSA-20...      -0.615926   \n",
      "206696  Regulation Of Gene Expression In Beta Cells R-...       0.615030   \n",
      "206707  NOTCH4 Intracellular Domain Regulates Transcri...       0.606829   \n",
      "206753              Elastic Fibre Formation R-HSA-1566948       0.586631   \n",
      "\n",
      "        nes score   p value  shuffled  \n",
      "0       -2.589219  0.009619     False  \n",
      "2       -2.421221  0.015468     False  \n",
      "4       -2.376529  0.017476     False  \n",
      "8        2.262510  0.023666     False  \n",
      "16       2.002924  0.045185     False  \n",
      "...           ...       ...       ...  \n",
      "206683   1.366001  0.171939     False  \n",
      "206690  -1.336073  0.181525     False  \n",
      "206696   1.303285  0.192478     False  \n",
      "206707   1.264243  0.206143     False  \n",
      "206753   1.168074  0.242777     False  \n",
      "\n",
      "[12386 rows x 8 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[I 2024-09-20 12:24:41,401] Trial 0 finished with value: 37.74495248367537 and parameters: {'beta': 1.9277921012058044, 'learning_rate': 7.776075823740862e-05, 'batch_size': 61, 'epochs': 250, 'optimizer_type': 'rmsprop'}. Best is trial 0 with value: 37.74495248367537.\n",
      "[I 2024-09-20 12:25:15,570] Trial 1 finished with value: 48.99225673390858 and parameters: {'beta': 1.0462210055019918, 'learning_rate': 3.945288877182508e-05, 'batch_size': 61, 'epochs': 407, 'optimizer_type': 'rmsprop'}. Best is trial 0 with value: 37.74495248367537.\n",
      "[I 2024-09-20 12:25:25,142] Trial 2 finished with value: 41.39392180940998 and parameters: {'beta': 1.6630818653084516, 'learning_rate': 9.556825880116263e-05, 'batch_size': 55, 'epochs': 111, 'optimizer_type': 'rmsprop'}. Best is trial 0 with value: 37.74495248367537.\n",
      "[I 2024-09-20 12:27:16,565] Trial 3 finished with value: 39.06326047982743 and parameters: {'beta': 1.7220471187673403, 'learning_rate': 5.462179749812174e-05, 'batch_size': 76, 'epochs': 936, 'optimizer_type': 'adam'}. Best is trial 0 with value: 37.74495248367537.\n",
      "[I 2024-09-20 12:28:08,229] Trial 4 finished with value: 42.87653307558885 and parameters: {'beta': 1.6953521919034655, 'learning_rate': 2.1472441498848508e-05, 'batch_size': 93, 'epochs': 451, 'optimizer_type': 'adam'}. Best is trial 0 with value: 37.74495248367537.\n",
      "[I 2024-09-20 12:28:52,793] Trial 5 finished with value: 44.76752219128964 and parameters: {'beta': 1.2353149778129544, 'learning_rate': 3.313354929130477e-05, 'batch_size': 45, 'epochs': 448, 'optimizer_type': 'adam'}. Best is trial 0 with value: 37.74495248367537.\n",
      "[I 2024-09-20 12:29:49,499] Trial 6 finished with value: 34.94089364578475 and parameters: {'beta': 1.7876941735059688, 'learning_rate': 4.48597764860786e-05, 'batch_size': 59, 'epochs': 704, 'optimizer_type': 'rmsprop'}. Best is trial 6 with value: 34.94089364578475.\n",
      "[I 2024-09-20 12:30:42,860] Trial 7 finished with value: 34.52432542487757 and parameters: {'beta': 1.847126841984522, 'learning_rate': 7.863550785477269e-05, 'batch_size': 50, 'epochs': 517, 'optimizer_type': 'adam'}. Best is trial 7 with value: 34.52432542487757.\n",
      "[I 2024-09-20 12:31:31,570] Trial 8 finished with value: 39.09286744986007 and parameters: {'beta': 1.9827548401629143, 'learning_rate': 8.151950283424502e-05, 'batch_size': 84, 'epochs': 437, 'optimizer_type': 'rmsprop'}. Best is trial 7 with value: 34.52432542487757.\n",
      "[I 2024-09-20 12:32:08,048] Trial 9 finished with value: 45.653030623250935 and parameters: {'beta': 1.3747214747189465, 'learning_rate': 8.498465636412222e-06, 'batch_size': 42, 'epochs': 338, 'optimizer_type': 'adam'}. Best is trial 7 with value: 34.52432542487757.\n",
      "[I 2024-09-20 12:34:06,456] Trial 10 finished with value: 29.428239964727144 and parameters: {'beta': 1.4929847373190286, 'learning_rate': 6.387785502977344e-05, 'batch_size': 20, 'epochs': 698, 'optimizer_type': 'adam'}. Best is trial 10 with value: 29.428239964727144.\n",
      "[I 2024-09-20 12:36:15,916] Trial 11 finished with value: 28.86604217984783 and parameters: {'beta': 1.4830943035922883, 'learning_rate': 6.376579387686104e-05, 'batch_size': 18, 'epochs': 681, 'optimizer_type': 'adam'}. Best is trial 11 with value: 28.86604217984783.\n",
      "[I 2024-09-20 12:38:23,453] Trial 12 finished with value: 28.874315335857336 and parameters: {'beta': 1.489601252740688, 'learning_rate': 6.045412199628242e-05, 'batch_size': 19, 'epochs': 713, 'optimizer_type': 'adam'}. Best is trial 11 with value: 28.86604217984783.\n",
      "[I 2024-09-20 12:40:53,432] Trial 13 finished with value: 27.58776486524895 and parameters: {'beta': 1.4564905758609628, 'learning_rate': 6.779541126914104e-05, 'batch_size': 16, 'epochs': 744, 'optimizer_type': 'adam'}. Best is trial 13 with value: 27.58776486524895.\n",
      "[I 2024-09-20 12:43:05,277] Trial 14 finished with value: 45.050339610541045 and parameters: {'beta': 1.3047100911072282, 'learning_rate': 6.73818213301104e-05, 'batch_size': 112, 'epochs': 923, 'optimizer_type': 'adam'}. Best is trial 13 with value: 27.58776486524895.\n",
      "[I 2024-09-20 12:44:52,461] Trial 15 finished with value: 30.782727870656483 and parameters: {'beta': 1.5764908732400515, 'learning_rate': 8.850970457062091e-05, 'batch_size': 31, 'epochs': 827, 'optimizer_type': 'adam'}. Best is trial 13 with value: 27.58776486524895.\n",
      "[I 2024-09-20 12:46:08,708] Trial 16 finished with value: 42.95206007316931 and parameters: {'beta': 1.0797582806696604, 'learning_rate': 6.898547398084339e-05, 'batch_size': 31, 'epochs': 595, 'optimizer_type': 'adam'}. Best is trial 13 with value: 27.58776486524895.\n",
      "[I 2024-09-20 12:47:53,996] Trial 17 finished with value: 35.7062950475892 and parameters: {'beta': 1.387441571793783, 'learning_rate': 5.420923558029316e-05, 'batch_size': 30, 'epochs': 793, 'optimizer_type': 'adam'}. Best is trial 13 with value: 27.58776486524895.\n",
      "[I 2024-09-20 12:50:05,035] Trial 18 finished with value: 34.76775744993295 and parameters: {'beta': 1.1936996326823612, 'learning_rate': 9.920028363907273e-05, 'batch_size': 17, 'epochs': 645, 'optimizer_type': 'adam'}. Best is trial 13 with value: 27.58776486524895.\n",
      "[I 2024-09-20 12:51:37,968] Trial 19 finished with value: 36.96513230053347 and parameters: {'beta': 1.587816425730094, 'learning_rate': 3.157106214187426e-05, 'batch_size': 41, 'epochs': 821, 'optimizer_type': 'adam'}. Best is trial 13 with value: 27.58776486524895.\n",
      "[I 2024-09-20 12:54:06,438] Trial 20 finished with value: 33.37880513205457 and parameters: {'beta': 1.4155244444791342, 'learning_rate': 4.9592818229164384e-05, 'batch_size': 26, 'epochs': 1000, 'optimizer_type': 'adam'}. Best is trial 13 with value: 27.58776486524895.\n",
      "[I 2024-09-20 12:56:02,927] Trial 21 finished with value: 26.28911155586812 and parameters: {'beta': 1.5139093633866547, 'learning_rate': 6.051111863587085e-05, 'batch_size': 16, 'epochs': 573, 'optimizer_type': 'adam'}. Best is trial 21 with value: 26.28911155586812.\n",
      "[I 2024-09-20 12:57:52,409] Trial 22 finished with value: 24.10234529865322 and parameters: {'beta': 1.5764144462377576, 'learning_rate': 7.298928426616051e-05, 'batch_size': 16, 'epochs': 540, 'optimizer_type': 'adam'}. Best is trial 22 with value: 24.10234529865322.\n",
      "[I 2024-09-20 12:59:03,933] Trial 23 finished with value: 31.89804047541832 and parameters: {'beta': 1.6281214961684585, 'learning_rate': 7.427664893909031e-05, 'batch_size': 35, 'epochs': 582, 'optimizer_type': 'adam'}. Best is trial 22 with value: 24.10234529865322.\n",
      "[I 2024-09-20 13:00:29,355] Trial 24 finished with value: 29.26719642753032 and parameters: {'beta': 1.5805290945619082, 'learning_rate': 8.561664963125783e-05, 'batch_size': 25, 'epochs': 548, 'optimizer_type': 'adam'}. Best is trial 22 with value: 24.10234529865322.\n",
      "[I 2024-09-20 13:01:14,305] Trial 25 finished with value: 42.563126521324044 and parameters: {'beta': 1.2913892841653145, 'learning_rate': 7.25851111973475e-05, 'batch_size': 37, 'epochs': 331, 'optimizer_type': 'adam'}. Best is trial 22 with value: 24.10234529865322.\n",
      "[I 2024-09-20 13:04:33,738] Trial 26 finished with value: 27.636083938826378 and parameters: {'beta': 1.4225095564482868, 'learning_rate': 5.850690654231348e-05, 'batch_size': 16, 'epochs': 780, 'optimizer_type': 'rmsprop'}. Best is trial 22 with value: 24.10234529865322.\n",
      "[I 2024-09-20 13:04:55,356] Trial 27 finished with value: 40.3049392016966 and parameters: {'beta': 1.7615214666132544, 'learning_rate': 8.836629526984472e-05, 'batch_size': 73, 'epochs': 184, 'optimizer_type': 'adam'}. Best is trial 22 with value: 24.10234529865322.\n",
      "[I 2024-09-20 13:06:52,176] Trial 28 finished with value: 28.99665049083197 and parameters: {'beta': 1.5477326317417026, 'learning_rate': 4.834494730880582e-05, 'batch_size': 22, 'epochs': 621, 'optimizer_type': 'adam'}. Best is trial 22 with value: 24.10234529865322.\n",
      "[I 2024-09-20 13:07:29,181] Trial 29 finished with value: 43.90793165520056 and parameters: {'beta': 1.3237742578748644, 'learning_rate': 7.51553868199275e-05, 'batch_size': 49, 'epochs': 307, 'optimizer_type': 'rmsprop'}. Best is trial 22 with value: 24.10234529865322.\n",
      "[I 2024-09-20 13:08:50,579] Trial 30 finished with value: 23.91354957694438 and parameters: {'beta': 1.8227118929652524, 'learning_rate': 8.194992936930852e-05, 'batch_size': 26, 'epochs': 504, 'optimizer_type': 'adam'}. Best is trial 30 with value: 23.91354957694438.\n",
      "[I 2024-09-20 13:10:12,286] Trial 31 finished with value: 20.74477821463969 and parameters: {'beta': 1.9341874703743978, 'learning_rate': 8.208845578394926e-05, 'batch_size': 25, 'epochs': 499, 'optimizer_type': 'adam'}. Best is trial 31 with value: 20.74477821463969.\n",
      "[I 2024-09-20 13:11:35,294] Trial 32 finished with value: 21.93952536511777 and parameters: {'beta': 1.8913264347083305, 'learning_rate': 9.218526047356133e-05, 'batch_size': 26, 'epochs': 529, 'optimizer_type': 'adam'}. Best is trial 31 with value: 20.74477821463969.\n",
      "[I 2024-09-20 13:12:33,038] Trial 33 finished with value: 23.95043362290112 and parameters: {'beta': 1.8963749215285144, 'learning_rate': 8.913013986260137e-05, 'batch_size': 27, 'epochs': 383, 'optimizer_type': 'adam'}. Best is trial 31 with value: 20.74477821463969.\n",
      "[I 2024-09-20 13:13:27,237] Trial 34 finished with value: 23.251559311596317 and parameters: {'beta': 1.8936843306669515, 'learning_rate': 9.338775333596187e-05, 'batch_size': 27, 'epochs': 388, 'optimizer_type': 'rmsprop'}. Best is trial 31 with value: 20.74477821463969.\n",
      "[I 2024-09-20 13:13:53,933] Trial 35 finished with value: 27.606247790891732 and parameters: {'beta': 1.993956764540643, 'learning_rate': 9.347418279283812e-05, 'batch_size': 35, 'epochs': 236, 'optimizer_type': 'rmsprop'}. Best is trial 31 with value: 20.74477821463969.\n",
      "[I 2024-09-20 13:13:58,003] Trial 36 finished with value: 38.395458073402516 and parameters: {'beta': 1.8721239646242183, 'learning_rate': 9.891259153622508e-05, 'batch_size': 38, 'epochs': 36, 'optimizer_type': 'rmsprop'}. Best is trial 31 with value: 20.74477821463969.\n",
      "[I 2024-09-20 13:14:44,616] Trial 37 finished with value: 35.256157445195896 and parameters: {'beta': 1.8318367339231665, 'learning_rate': 8.145667709984965e-05, 'batch_size': 53, 'epochs': 462, 'optimizer_type': 'rmsprop'}. Best is trial 31 with value: 20.74477821463969.\n",
      "[I 2024-09-20 13:15:33,222] Trial 38 finished with value: 37.507518620277516 and parameters: {'beta': 1.9375182709475713, 'learning_rate': 9.223987025203699e-05, 'batch_size': 70, 'epochs': 487, 'optimizer_type': 'rmsprop'}. Best is trial 31 with value: 20.74477821463969.\n",
      "[I 2024-09-20 13:16:10,816] Trial 39 finished with value: 35.60943157139109 and parameters: {'beta': 1.7805568338782072, 'learning_rate': 8.42673027769113e-05, 'batch_size': 59, 'epochs': 409, 'optimizer_type': 'rmsprop'}. Best is trial 31 with value: 20.74477821463969.\n",
      "[I 2024-09-20 13:16:36,567] Trial 40 finished with value: 32.927984937980995 and parameters: {'beta': 1.928474593483658, 'learning_rate': 9.53451127793843e-05, 'batch_size': 47, 'epochs': 260, 'optimizer_type': 'rmsprop'}. Best is trial 31 with value: 20.74477821463969.\n",
      "[I 2024-09-20 13:17:32,823] Trial 41 finished with value: 22.347145285535216 and parameters: {'beta': 1.8859339794151797, 'learning_rate': 8.975619882825433e-05, 'batch_size': 26, 'epochs': 365, 'optimizer_type': 'adam'}. Best is trial 31 with value: 20.74477821463969.\n",
      "[I 2024-09-20 13:18:37,138] Trial 42 finished with value: 22.673982739804394 and parameters: {'beta': 1.810648910243131, 'learning_rate': 7.938094817615707e-05, 'batch_size': 23, 'epochs': 381, 'optimizer_type': 'adam'}. Best is trial 31 with value: 20.74477821463969.\n",
      "[I 2024-09-20 13:19:26,764] Trial 43 finished with value: 32.16514710383629 and parameters: {'beta': 1.7309229400895478, 'learning_rate': 9.153824378176171e-05, 'batch_size': 33, 'epochs': 376, 'optimizer_type': 'adam'}. Best is trial 31 with value: 20.74477821463969.\n",
      "[I 2024-09-20 13:20:06,793] Trial 44 finished with value: 21.907250338881763 and parameters: {'beta': 1.9433711999634544, 'learning_rate': 7.911348789878644e-05, 'batch_size': 24, 'epochs': 282, 'optimizer_type': 'rmsprop'}. Best is trial 31 with value: 20.74477821463969.\n",
      "[I 2024-09-20 13:20:41,327] Trial 45 finished with value: 22.200943448650303 and parameters: {'beta': 1.956749345791957, 'learning_rate': 7.741966487011449e-05, 'batch_size': 22, 'epochs': 197, 'optimizer_type': 'adam'}. Best is trial 31 with value: 20.74477821463969.\n",
      "[I 2024-09-20 13:20:59,525] Trial 46 finished with value: 32.954028138118005 and parameters: {'beta': 1.9537962767545605, 'learning_rate': 7.71048195766116e-05, 'batch_size': 43, 'epochs': 155, 'optimizer_type': 'adam'}. Best is trial 31 with value: 20.74477821463969.\n",
      "[I 2024-09-20 13:21:08,780] Trial 47 finished with value: 74.78581506529851 and parameters: {'beta': 1.679513587913307, 'learning_rate': 6.89709344851354e-06, 'batch_size': 89, 'epochs': 77, 'optimizer_type': 'adam'}. Best is trial 31 with value: 20.74477821463969.\n",
      "[I 2024-09-20 13:21:50,852] Trial 48 finished with value: 18.161248141616138 and parameters: {'beta': 1.9977987530171926, 'learning_rate': 8.550261486325946e-05, 'batch_size': 22, 'epochs': 265, 'optimizer_type': 'rmsprop'}. Best is trial 48 with value: 18.161248141616138.\n",
      "[I 2024-09-20 13:22:15,943] Trial 49 finished with value: 38.156391155185986 and parameters: {'beta': 1.974254913507335, 'learning_rate': 2.0549683551676e-05, 'batch_size': 65, 'epochs': 260, 'optimizer_type': 'rmsprop'}. Best is trial 48 with value: 18.161248141616138.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running GSEA for betatcvae\n",
      "          z  full_model_z      model  \\\n",
      "0         1           200  betatcvae   \n",
      "1         1           200  betatcvae   \n",
      "3         1           200  betatcvae   \n",
      "8         1           200  betatcvae   \n",
      "9         1           200  betatcvae   \n",
      "...     ...           ...        ...   \n",
      "206768  200           200  betatcvae   \n",
      "206769  200           200  betatcvae   \n",
      "206770  200           200  betatcvae   \n",
      "206773  200           200  betatcvae   \n",
      "206774  200           200  betatcvae   \n",
      "\n",
      "                                         reactome pathway  gsea es score  \\\n",
      "0       Regulation Of TP53 Activity Thru Acetylation R...      -0.638290   \n",
      "1       STAT3 Nuclear Events Downstream Of ALK Signali...      -0.793609   \n",
      "3       SUMOylation Of Transcription Factors R-HSA-323...       0.674916   \n",
      "8       Downregulation Of TGF-beta Receptor Signaling ...       0.619637   \n",
      "9       Regulation Of IGF Transport And Uptake By IGFB...      -0.848614   \n",
      "...                                                   ...            ...   \n",
      "206768  Non-integrin membrane-ECM Interactions R-HSA-3...      -0.595919   \n",
      "206769   Metabolism Of Fat-Soluble Vitamins R-HSA-6806667      -0.595701   \n",
      "206770        PINK1-PRKN Mediated Mitophagy R-HSA-5205685       0.596022   \n",
      "206773           Dimerization Of Procaspase-8 R-HSA-69416      -0.592925   \n",
      "206774          CASP8 Activity Is Inhibited R-HSA-5218900      -0.592925   \n",
      "\n",
      "        nes score   p value  shuffled  \n",
      "0       -2.792424  0.005231     False  \n",
      "1       -2.656548  0.007895     False  \n",
      "3        2.648964  0.008074     False  \n",
      "8        2.489596  0.012789     False  \n",
      "9       -2.484994  0.012955     False  \n",
      "...           ...       ...       ...  \n",
      "206768  -1.461029  0.144008     False  \n",
      "206769  -1.459924  0.144311     False  \n",
      "206770   1.455898  0.145421     False  \n",
      "206773  -1.445840  0.148222     False  \n",
      "206774  -1.445840  0.148222     False  \n",
      "\n",
      "[8124 rows x 8 columns]\n"
     ]
    }
   ],
   "source": [
    "def extract_weights(model, model_name, H=None):\n",
    "    if model_name in [\"pca\", \"ica\", \"nmf\"]:\n",
    "        # Transpose PCA components and format columns\n",
    "        weights_df = pd.DataFrame(model.components_, columns=dependency_df.drop(columns=[\"ModelID\"]).columns.tolist()).transpose()\n",
    "        weights_df.columns = [f\"{x}\" for x in range(0, weights_df.shape[1])]\n",
    "        \n",
    "    # Reset index to rename 'index' to 'genes'\n",
    "    weights_df = weights_df.reset_index().rename(columns={\"index\": \"genes\"})\n",
    "    \n",
    "    return weights_df\n",
    "\n",
    "\n",
    "def perform_gsea(weights_df, model_name, num_components):\n",
    "    library = blitz.enrichr.get_library(\"Reactome_2022\")\n",
    "    seed = random.random()\n",
    "    gsea_results = []\n",
    "    # Define cutoff values\n",
    "    lfc_cutoff = 0.584\n",
    "    fdr_cutoff = 0.25\n",
    "    # Perform GSEA for each component column in weights_df\n",
    "    for col in weights_df.columns[1:]:  # Skip 'genes' column\n",
    "        gene_signature = weights_df[['genes', col]]\n",
    "        \n",
    "        if gene_signature.shape[0] > 0:\n",
    "            try:\n",
    "                # Perform GSEA using the gene signature (weights)\n",
    "                gsea_result = blitz.gsea(gene_signature, library, seed=seed)\n",
    "                gsea_result = gsea_result.reset_index()\n",
    "                for _, pathway_result in gsea_result.iterrows():\n",
    "                    result_row = {\n",
    "                        \"z\": int(col),\n",
    "                        \"full_model_z\": num_components,\n",
    "                        \"model\": str(model_name),\n",
    "                        \"reactome pathway\": str(pathway_result['Term']),\n",
    "                        \"gsea es score\": pathway_result['es'],\n",
    "                        \"nes score\": pathway_result['nes'],\n",
    "                        \"p value\": pathway_result['pval'],\n",
    "                        \"shuffled\": False\n",
    "                    }\n",
    "                    gsea_results.append(result_row)\n",
    "                    \n",
    "            except ZeroDivisionError:\n",
    "                print(f\"Skipping GSEA for {col} due to zero division error.\")\n",
    "    \n",
    "    # Convert GSEA results into a DataFrame and save to a Parquet file\n",
    "    gsea_results_df = pd.DataFrame(gsea_results)\n",
    "    # Filter for significant results\n",
    "    significant_gsea_df = gsea_results_df[\n",
    "            (gsea_results_df['gsea es score'].abs() > lfc_cutoff) &\n",
    "            (gsea_results_df['p value'] < fdr_cutoff)\n",
    "        ]\n",
    "    print(significant_gsea_df)\n",
    "    \n",
    "    return significant_gsea_df\n",
    "\n",
    "\n",
    "# Initialize a list to store the final results\n",
    "final_gsea_results = []\n",
    "\n",
    "# GSEA settings\n",
    "latent_dims = [2, 3, 4, 5, 6, 7, 8, 9, 10, 12, 14, 16, 18, 20, 25, 30, 35, 40, 45, 50, 60, 70, 80, 90, 100, 150, 200]\n",
    "model_names = [\"pca\", \"ica\", \"nmf\", \"vanillavae\", \"betavae\", \"betatcvae\"]\n",
    "\n",
    "# Define the output file path\n",
    "final_output_file = pathlib.Path(output_dir) / \"combined_z_matrix_gsea_results.parquet\"\n",
    "\n",
    "# Try to load the existing combined results DataFrame if it exists\n",
    "try:\n",
    "    combined_results_df = pd.read_parquet(final_output_file)\n",
    "    print(f\"Loaded existing combined results from {final_output_file}\")\n",
    "except FileNotFoundError:\n",
    "    # If the file doesn't exist, initialize an empty DataFrame\n",
    "    combined_results_df = pd.DataFrame()\n",
    "    print(f\"No existing file found. Initialized empty DataFrame.\")\n",
    "\n",
    "for num_components in latent_dims:\n",
    "    for model_name in model_names:\n",
    "        # Check if this model and latent dimension have already been processed\n",
    "        if not combined_results_df.empty:\n",
    "            if ((combined_results_df['model'] == model_name) & \n",
    "                (combined_results_df['full_model_z'] == num_components)).any():\n",
    "                print(f\"Skipping {model_name} with {num_components} dimensions as it is already processed.\")\n",
    "                continue  # Skip to the next iteration if this combination is already present\n",
    "        z_matrix_df = None\n",
    "        \n",
    "        if model_name in [\"pca\", \"ica\", \"nmf\"]:\n",
    "            # Sklearn models (PCA, ICA, NMF)\n",
    "            if model_name == \"pca\":\n",
    "                model = PCA(n_components=num_components)\n",
    "            elif model_name == \"ica\":\n",
    "                model = FastICA(n_components=num_components)\n",
    "            elif model_name == \"nmf\":\n",
    "                model = NMF(n_components=num_components, init='nndsvd', max_iter=1000, random_state=0)\n",
    "            \n",
    "            model.fit(train_data)\n",
    "            H = model.transform(train_data) if model_name == \"nmf\" else None\n",
    "            weight_matrix_df = extract_weights(model, model_name, H)\n",
    "                \n",
    "        elif model_name == \"betavae\":\n",
    "            # Optuna optimization for BetaVAE\n",
    "            study = optuna.create_study(direction=\"minimize\")\n",
    "            study.optimize(lambda trial: objective(trial, train_tensor, train_tensor, train_data, latent_dim=num_components), n_trials=50)\n",
    "            \n",
    "            # Train the best BetaVAE model and extract z matrix\n",
    "            best_trial = study.best_trial\n",
    "            model = BetaVAE(input_dim=train_data.shape[1], latent_dim=num_components, beta=best_trial.params['beta'])\n",
    "            train_loader = DataLoader(TensorDataset(train_tensor), batch_size=best_trial.params['batch_size'], shuffle=True)\n",
    "            optimizer = get_optimizer(best_trial.params['optimizer_type'], model.parameters(), best_trial.params['learning_rate'])\n",
    "            train_vae(model, train_loader, optimizer, best_trial.params['epochs'])\n",
    "            \n",
    "            weight_matrix_df = weights(model, weight_data)\n",
    "            weight_matrix_df.rename(columns={0: 'genes'}, inplace=True)\n",
    "        elif model_name == \"betatcvae\":\n",
    "            # Optuna optimization for BetaTCVAE\n",
    "            study = optuna.create_study(direction=\"minimize\")\n",
    "            study.optimize(lambda trial: objective_tc(trial, train_tensor, train_tensor, train_data, latent_dim=num_components), n_trials=50)\n",
    "            # Train the best BetaTCVAE model and extract z matrix\n",
    "            best_trial = study.best_trial\n",
    "            model = BetaTCVAE(input_dim=train_data.shape[1], latent_dim=num_components, beta=best_trial.params['beta'])\n",
    "            train_loader = DataLoader(TensorDataset(train_tensor), batch_size=best_trial.params['batch_size'], shuffle=True)\n",
    "            optimizer = get_optimizer_tc(best_trial.params['optimizer_type'], model.parameters(), best_trial.params['learning_rate'])\n",
    "            train_tc_vae(model, train_loader, optimizer, best_trial.params['epochs'])\n",
    "            \n",
    "            # Extract weight matrix\n",
    "            weight_matrix_df = tc_weights(model, weight_data)\n",
    "            weight_matrix_df.rename(columns={0: 'genes'}, inplace=True)\n",
    "        \n",
    "        elif model_name == \"vanillavae\":\n",
    "            # Optuna optimization for VanillaVAE\n",
    "            study = optuna.create_study(direction=\"minimize\")\n",
    "            study.optimize(lambda trial: objective_vvae(trial, train_tensor, train_tensor, train_data, latent_dim=num_components), n_trials=50)\n",
    "            \n",
    "            # Train the best BetaTCVAE model and extract z matrix\n",
    "            best_trial = study.best_trial\n",
    "            model = VanillaVAE(input_dim=train_data.shape[1], latent_dim=num_components)\n",
    "            train_loader = DataLoader(TensorDataset(train_tensor), batch_size=best_trial.params['batch_size'], shuffle=True)\n",
    "            optimizer = get_optimizer_vvae(best_trial.params['optimizer_type'], model.parameters(), best_trial.params['learning_rate'])\n",
    "            train_vvae(model, train_loader, optimizer, best_trial.params['epochs'])\n",
    "            \n",
    "            # Extract weight matrix\n",
    "            weight_matrix_df = vanilla_weights(model, weight_data)\n",
    "            weight_matrix_df.rename(columns={0: 'genes'}, inplace=True)\n",
    "        # If weight_matrix is generated, proceed to GSEA and append to combined dataframe\n",
    "        if weight_matrix_df is not None:\n",
    "            print(f\"Running GSEA for {model_name}\")\n",
    "            gsea_results_df = perform_gsea(weight_matrix_df, model_name, num_components)\n",
    "            combined_results_df = pd.concat([combined_results_df, gsea_results_df], ignore_index=True)\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved final combined z_matrix and GSEA results to /home/juliacurd/gene_dependency_representations/4.gene_expression_signatures/results/combined_z_matrix_gsea_results.parquet\n"
     ]
    }
   ],
   "source": [
    "# Save the combined dataframe to a file\n",
    "final_output_file = output_dir / \"combined_z_matrix_gsea_results.parquet\"\n",
    "combined_results_df.to_parquet(final_output_file, index=False)\n",
    "\n",
    "print(f\"Saved final combined z_matrix and GSEA results to {final_output_file}\")\n",
    "\n",
    "#Save as CSV for R \n",
    "csv_output_file = output_dir / \"combined_z_matrix_gsea_results.csv\"\n",
    "combined_results_df.to_csv(csv_output_file, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>z</th>\n",
       "      <th>full_model_z</th>\n",
       "      <th>model</th>\n",
       "      <th>reactome pathway</th>\n",
       "      <th>gsea es score</th>\n",
       "      <th>nes score</th>\n",
       "      <th>p value</th>\n",
       "      <th>shuffled</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>94309</th>\n",
       "      <td>21</td>\n",
       "      <td>40</td>\n",
       "      <td>betavae</td>\n",
       "      <td>Regulation Of IGF Transport And Uptake By IGFB...</td>\n",
       "      <td>0.982719</td>\n",
       "      <td>2.749145</td>\n",
       "      <td>0.005975</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94308</th>\n",
       "      <td>21</td>\n",
       "      <td>40</td>\n",
       "      <td>betavae</td>\n",
       "      <td>Post-translational Protein Phosphorylation R-H...</td>\n",
       "      <td>0.982719</td>\n",
       "      <td>2.749145</td>\n",
       "      <td>0.005975</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28586</th>\n",
       "      <td>5</td>\n",
       "      <td>14</td>\n",
       "      <td>vanillavae</td>\n",
       "      <td>Post-translational Protein Phosphorylation R-H...</td>\n",
       "      <td>0.976058</td>\n",
       "      <td>2.868438</td>\n",
       "      <td>0.004125</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28587</th>\n",
       "      <td>5</td>\n",
       "      <td>14</td>\n",
       "      <td>vanillavae</td>\n",
       "      <td>Regulation Of IGF Transport And Uptake By IGFB...</td>\n",
       "      <td>0.976058</td>\n",
       "      <td>2.868438</td>\n",
       "      <td>0.004125</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174148</th>\n",
       "      <td>26</td>\n",
       "      <td>70</td>\n",
       "      <td>betavae</td>\n",
       "      <td>Post-translational Protein Phosphorylation R-H...</td>\n",
       "      <td>-0.974218</td>\n",
       "      <td>-2.913465</td>\n",
       "      <td>0.003574</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174149</th>\n",
       "      <td>26</td>\n",
       "      <td>70</td>\n",
       "      <td>betavae</td>\n",
       "      <td>Regulation Of IGF Transport And Uptake By IGFB...</td>\n",
       "      <td>-0.974218</td>\n",
       "      <td>-2.913465</td>\n",
       "      <td>0.003574</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176386</th>\n",
       "      <td>62</td>\n",
       "      <td>70</td>\n",
       "      <td>betavae</td>\n",
       "      <td>Post-translational Protein Phosphorylation R-H...</td>\n",
       "      <td>-0.970980</td>\n",
       "      <td>-3.084574</td>\n",
       "      <td>0.002038</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176387</th>\n",
       "      <td>62</td>\n",
       "      <td>70</td>\n",
       "      <td>betavae</td>\n",
       "      <td>Regulation Of IGF Transport And Uptake By IGFB...</td>\n",
       "      <td>-0.970980</td>\n",
       "      <td>-3.084574</td>\n",
       "      <td>0.002038</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222392</th>\n",
       "      <td>2</td>\n",
       "      <td>90</td>\n",
       "      <td>nmf</td>\n",
       "      <td>Viral mRNA Translation R-HSA-192823</td>\n",
       "      <td>0.968566</td>\n",
       "      <td>3.166789</td>\n",
       "      <td>0.001541</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56865</th>\n",
       "      <td>6</td>\n",
       "      <td>25</td>\n",
       "      <td>betavae</td>\n",
       "      <td>Regulation Of IGF Transport And Uptake By IGFB...</td>\n",
       "      <td>0.963643</td>\n",
       "      <td>2.771649</td>\n",
       "      <td>0.005577</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56864</th>\n",
       "      <td>6</td>\n",
       "      <td>25</td>\n",
       "      <td>betavae</td>\n",
       "      <td>Post-translational Protein Phosphorylation R-H...</td>\n",
       "      <td>0.963643</td>\n",
       "      <td>2.771649</td>\n",
       "      <td>0.005577</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158480</th>\n",
       "      <td>61</td>\n",
       "      <td>70</td>\n",
       "      <td>pca</td>\n",
       "      <td>DCC Mediated Attractive Signaling R-HSA-418885</td>\n",
       "      <td>-0.960806</td>\n",
       "      <td>-2.808619</td>\n",
       "      <td>0.004975</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233377</th>\n",
       "      <td>5</td>\n",
       "      <td>90</td>\n",
       "      <td>betavae</td>\n",
       "      <td>Regulation Of IGF Transport And Uptake By IGFB...</td>\n",
       "      <td>0.959894</td>\n",
       "      <td>2.732804</td>\n",
       "      <td>0.006280</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233376</th>\n",
       "      <td>5</td>\n",
       "      <td>90</td>\n",
       "      <td>betavae</td>\n",
       "      <td>Post-translational Protein Phosphorylation R-H...</td>\n",
       "      <td>0.959894</td>\n",
       "      <td>2.732804</td>\n",
       "      <td>0.006280</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230340</th>\n",
       "      <td>48</td>\n",
       "      <td>90</td>\n",
       "      <td>vanillavae</td>\n",
       "      <td>Uptake And Function Of Anthrax Toxins R-HSA-52...</td>\n",
       "      <td>-0.959825</td>\n",
       "      <td>-3.091930</td>\n",
       "      <td>0.001989</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61856</th>\n",
       "      <td>9</td>\n",
       "      <td>30</td>\n",
       "      <td>ica</td>\n",
       "      <td>Detoxification Of Reactive Oxygen Species R-HS...</td>\n",
       "      <td>-0.958717</td>\n",
       "      <td>-2.924281</td>\n",
       "      <td>0.003453</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22327</th>\n",
       "      <td>7</td>\n",
       "      <td>12</td>\n",
       "      <td>ica</td>\n",
       "      <td>Regulation Of IGF Transport And Uptake By IGFB...</td>\n",
       "      <td>-0.957895</td>\n",
       "      <td>-2.850276</td>\n",
       "      <td>0.004368</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22326</th>\n",
       "      <td>7</td>\n",
       "      <td>12</td>\n",
       "      <td>ica</td>\n",
       "      <td>Post-translational Protein Phosphorylation R-H...</td>\n",
       "      <td>-0.957895</td>\n",
       "      <td>-2.850276</td>\n",
       "      <td>0.004368</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27385</th>\n",
       "      <td>12</td>\n",
       "      <td>14</td>\n",
       "      <td>ica</td>\n",
       "      <td>Regulation Of IGF Transport And Uptake By IGFB...</td>\n",
       "      <td>0.957042</td>\n",
       "      <td>2.705392</td>\n",
       "      <td>0.006822</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27384</th>\n",
       "      <td>12</td>\n",
       "      <td>14</td>\n",
       "      <td>ica</td>\n",
       "      <td>Post-translational Protein Phosphorylation R-H...</td>\n",
       "      <td>0.957042</td>\n",
       "      <td>2.705392</td>\n",
       "      <td>0.006822</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68476</th>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>betavae</td>\n",
       "      <td>Regulation Of IGF Transport And Uptake By IGFB...</td>\n",
       "      <td>-0.956472</td>\n",
       "      <td>-2.838184</td>\n",
       "      <td>0.004537</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68477</th>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>betavae</td>\n",
       "      <td>Post-translational Protein Phosphorylation R-H...</td>\n",
       "      <td>-0.956472</td>\n",
       "      <td>-2.838184</td>\n",
       "      <td>0.004537</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143780</th>\n",
       "      <td>53</td>\n",
       "      <td>60</td>\n",
       "      <td>nmf</td>\n",
       "      <td>Detoxification Of Reactive Oxygen Species R-HS...</td>\n",
       "      <td>0.955140</td>\n",
       "      <td>3.238024</td>\n",
       "      <td>0.001204</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>371855</th>\n",
       "      <td>150</td>\n",
       "      <td>200</td>\n",
       "      <td>nmf</td>\n",
       "      <td>Synthesis Of PC R-HSA-1483191</td>\n",
       "      <td>0.954797</td>\n",
       "      <td>2.862941</td>\n",
       "      <td>0.004197</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175058</th>\n",
       "      <td>41</td>\n",
       "      <td>70</td>\n",
       "      <td>betavae</td>\n",
       "      <td>Synthesis Of PC R-HSA-1483191</td>\n",
       "      <td>-0.954603</td>\n",
       "      <td>-2.827130</td>\n",
       "      <td>0.004697</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39844</th>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "      <td>vanillavae</td>\n",
       "      <td>Post-translational Protein Phosphorylation R-H...</td>\n",
       "      <td>0.954468</td>\n",
       "      <td>2.770181</td>\n",
       "      <td>0.005603</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39843</th>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "      <td>vanillavae</td>\n",
       "      <td>Regulation Of IGF Transport And Uptake By IGFB...</td>\n",
       "      <td>0.954468</td>\n",
       "      <td>2.770181</td>\n",
       "      <td>0.005603</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>377304</th>\n",
       "      <td>44</td>\n",
       "      <td>200</td>\n",
       "      <td>vanillavae</td>\n",
       "      <td>Regulation Of IFNG Signaling R-HSA-877312</td>\n",
       "      <td>-0.954211</td>\n",
       "      <td>-2.816713</td>\n",
       "      <td>0.004852</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>319609</th>\n",
       "      <td>24</td>\n",
       "      <td>150</td>\n",
       "      <td>betavae</td>\n",
       "      <td>Post-translational Protein Phosphorylation R-H...</td>\n",
       "      <td>0.952745</td>\n",
       "      <td>2.791679</td>\n",
       "      <td>0.005244</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>319610</th>\n",
       "      <td>24</td>\n",
       "      <td>150</td>\n",
       "      <td>betavae</td>\n",
       "      <td>Regulation Of IGF Transport And Uptake By IGFB...</td>\n",
       "      <td>0.952745</td>\n",
       "      <td>2.791679</td>\n",
       "      <td>0.005244</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323086</th>\n",
       "      <td>80</td>\n",
       "      <td>150</td>\n",
       "      <td>betavae</td>\n",
       "      <td>Regulation Of IGF Transport And Uptake By IGFB...</td>\n",
       "      <td>-0.951999</td>\n",
       "      <td>-2.788009</td>\n",
       "      <td>0.005303</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323087</th>\n",
       "      <td>80</td>\n",
       "      <td>150</td>\n",
       "      <td>betavae</td>\n",
       "      <td>Post-translational Protein Phosphorylation R-H...</td>\n",
       "      <td>-0.951999</td>\n",
       "      <td>-2.788009</td>\n",
       "      <td>0.005303</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143474</th>\n",
       "      <td>47</td>\n",
       "      <td>60</td>\n",
       "      <td>nmf</td>\n",
       "      <td>rRNA Modification In Mitochondrion R-HSA-6793080</td>\n",
       "      <td>0.951905</td>\n",
       "      <td>3.251298</td>\n",
       "      <td>0.001149</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41113</th>\n",
       "      <td>4</td>\n",
       "      <td>18</td>\n",
       "      <td>betavae</td>\n",
       "      <td>Cytochrome P450 - Arranged By Substrate Type R...</td>\n",
       "      <td>-0.951814</td>\n",
       "      <td>-2.536189</td>\n",
       "      <td>0.011207</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>237219</th>\n",
       "      <td>67</td>\n",
       "      <td>90</td>\n",
       "      <td>betavae</td>\n",
       "      <td>Regulation Of IGF Transport And Uptake By IGFB...</td>\n",
       "      <td>0.949431</td>\n",
       "      <td>2.729118</td>\n",
       "      <td>0.006350</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>237220</th>\n",
       "      <td>67</td>\n",
       "      <td>90</td>\n",
       "      <td>betavae</td>\n",
       "      <td>Post-translational Protein Phosphorylation R-H...</td>\n",
       "      <td>0.949431</td>\n",
       "      <td>2.729118</td>\n",
       "      <td>0.006350</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91242</th>\n",
       "      <td>11</td>\n",
       "      <td>40</td>\n",
       "      <td>vanillavae</td>\n",
       "      <td>Post-translational Protein Phosphorylation R-H...</td>\n",
       "      <td>0.948840</td>\n",
       "      <td>2.847893</td>\n",
       "      <td>0.004401</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91243</th>\n",
       "      <td>11</td>\n",
       "      <td>40</td>\n",
       "      <td>vanillavae</td>\n",
       "      <td>Regulation Of IGF Transport And Uptake By IGFB...</td>\n",
       "      <td>0.948840</td>\n",
       "      <td>2.847893</td>\n",
       "      <td>0.004401</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>226405</th>\n",
       "      <td>73</td>\n",
       "      <td>90</td>\n",
       "      <td>nmf</td>\n",
       "      <td>Viral mRNA Translation R-HSA-192823</td>\n",
       "      <td>-0.947763</td>\n",
       "      <td>-3.035439</td>\n",
       "      <td>0.002402</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57659</th>\n",
       "      <td>18</td>\n",
       "      <td>25</td>\n",
       "      <td>betavae</td>\n",
       "      <td>Regulation Of IGF Transport And Uptake By IGFB...</td>\n",
       "      <td>0.946816</td>\n",
       "      <td>2.750558</td>\n",
       "      <td>0.005949</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57658</th>\n",
       "      <td>18</td>\n",
       "      <td>25</td>\n",
       "      <td>betavae</td>\n",
       "      <td>Post-translational Protein Phosphorylation R-H...</td>\n",
       "      <td>0.946816</td>\n",
       "      <td>2.750558</td>\n",
       "      <td>0.005949</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>363425</th>\n",
       "      <td>195</td>\n",
       "      <td>200</td>\n",
       "      <td>ica</td>\n",
       "      <td>IL-6-type Cytokine Receptor Ligand Interaction...</td>\n",
       "      <td>0.946068</td>\n",
       "      <td>2.706974</td>\n",
       "      <td>0.006790</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>319500</th>\n",
       "      <td>22</td>\n",
       "      <td>150</td>\n",
       "      <td>betavae</td>\n",
       "      <td>Regulation Of IGF Transport And Uptake By IGFB...</td>\n",
       "      <td>0.945640</td>\n",
       "      <td>2.733541</td>\n",
       "      <td>0.006266</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>319499</th>\n",
       "      <td>22</td>\n",
       "      <td>150</td>\n",
       "      <td>betavae</td>\n",
       "      <td>Post-translational Protein Phosphorylation R-H...</td>\n",
       "      <td>0.945640</td>\n",
       "      <td>2.733541</td>\n",
       "      <td>0.006266</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110989</th>\n",
       "      <td>36</td>\n",
       "      <td>45</td>\n",
       "      <td>betavae</td>\n",
       "      <td>Regulation Of IGF Transport And Uptake By IGFB...</td>\n",
       "      <td>0.944881</td>\n",
       "      <td>2.899886</td>\n",
       "      <td>0.003733</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110990</th>\n",
       "      <td>36</td>\n",
       "      <td>45</td>\n",
       "      <td>betavae</td>\n",
       "      <td>Post-translational Protein Phosphorylation R-H...</td>\n",
       "      <td>0.944881</td>\n",
       "      <td>2.899886</td>\n",
       "      <td>0.003733</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282261</th>\n",
       "      <td>50</td>\n",
       "      <td>150</td>\n",
       "      <td>pca</td>\n",
       "      <td>Receptor Mediated Mitophagy R-HSA-8934903</td>\n",
       "      <td>-0.943841</td>\n",
       "      <td>-2.766957</td>\n",
       "      <td>0.005658</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>375673</th>\n",
       "      <td>18</td>\n",
       "      <td>200</td>\n",
       "      <td>vanillavae</td>\n",
       "      <td>Regulation Of IGF Transport And Uptake By IGFB...</td>\n",
       "      <td>-0.943403</td>\n",
       "      <td>-2.779569</td>\n",
       "      <td>0.005443</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>375672</th>\n",
       "      <td>18</td>\n",
       "      <td>200</td>\n",
       "      <td>vanillavae</td>\n",
       "      <td>Post-translational Protein Phosphorylation R-H...</td>\n",
       "      <td>-0.943403</td>\n",
       "      <td>-2.779569</td>\n",
       "      <td>0.005443</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367555</th>\n",
       "      <td>64</td>\n",
       "      <td>200</td>\n",
       "      <td>nmf</td>\n",
       "      <td>Cytochrome P450 - Arranged By Substrate Type R...</td>\n",
       "      <td>0.941947</td>\n",
       "      <td>3.012471</td>\n",
       "      <td>0.002591</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          z  full_model_z       model  \\\n",
       "94309    21            40     betavae   \n",
       "94308    21            40     betavae   \n",
       "28586     5            14  vanillavae   \n",
       "28587     5            14  vanillavae   \n",
       "174148   26            70     betavae   \n",
       "174149   26            70     betavae   \n",
       "176386   62            70     betavae   \n",
       "176387   62            70     betavae   \n",
       "222392    2            90         nmf   \n",
       "56865     6            25     betavae   \n",
       "56864     6            25     betavae   \n",
       "158480   61            70         pca   \n",
       "233377    5            90     betavae   \n",
       "233376    5            90     betavae   \n",
       "230340   48            90  vanillavae   \n",
       "61856     9            30         ica   \n",
       "22327     7            12         ica   \n",
       "22326     7            12         ica   \n",
       "27385    12            14         ica   \n",
       "27384    12            14         ica   \n",
       "68476    30            30     betavae   \n",
       "68477    30            30     betavae   \n",
       "143780   53            60         nmf   \n",
       "371855  150           200         nmf   \n",
       "175058   41            70     betavae   \n",
       "39844     1            18  vanillavae   \n",
       "39843     1            18  vanillavae   \n",
       "377304   44           200  vanillavae   \n",
       "319609   24           150     betavae   \n",
       "319610   24           150     betavae   \n",
       "323086   80           150     betavae   \n",
       "323087   80           150     betavae   \n",
       "143474   47            60         nmf   \n",
       "41113     4            18     betavae   \n",
       "237219   67            90     betavae   \n",
       "237220   67            90     betavae   \n",
       "91242    11            40  vanillavae   \n",
       "91243    11            40  vanillavae   \n",
       "226405   73            90         nmf   \n",
       "57659    18            25     betavae   \n",
       "57658    18            25     betavae   \n",
       "363425  195           200         ica   \n",
       "319500   22           150     betavae   \n",
       "319499   22           150     betavae   \n",
       "110989   36            45     betavae   \n",
       "110990   36            45     betavae   \n",
       "282261   50           150         pca   \n",
       "375673   18           200  vanillavae   \n",
       "375672   18           200  vanillavae   \n",
       "367555   64           200         nmf   \n",
       "\n",
       "                                         reactome pathway  gsea es score  \\\n",
       "94309   Regulation Of IGF Transport And Uptake By IGFB...       0.982719   \n",
       "94308   Post-translational Protein Phosphorylation R-H...       0.982719   \n",
       "28586   Post-translational Protein Phosphorylation R-H...       0.976058   \n",
       "28587   Regulation Of IGF Transport And Uptake By IGFB...       0.976058   \n",
       "174148  Post-translational Protein Phosphorylation R-H...      -0.974218   \n",
       "174149  Regulation Of IGF Transport And Uptake By IGFB...      -0.974218   \n",
       "176386  Post-translational Protein Phosphorylation R-H...      -0.970980   \n",
       "176387  Regulation Of IGF Transport And Uptake By IGFB...      -0.970980   \n",
       "222392                Viral mRNA Translation R-HSA-192823       0.968566   \n",
       "56865   Regulation Of IGF Transport And Uptake By IGFB...       0.963643   \n",
       "56864   Post-translational Protein Phosphorylation R-H...       0.963643   \n",
       "158480     DCC Mediated Attractive Signaling R-HSA-418885      -0.960806   \n",
       "233377  Regulation Of IGF Transport And Uptake By IGFB...       0.959894   \n",
       "233376  Post-translational Protein Phosphorylation R-H...       0.959894   \n",
       "230340  Uptake And Function Of Anthrax Toxins R-HSA-52...      -0.959825   \n",
       "61856   Detoxification Of Reactive Oxygen Species R-HS...      -0.958717   \n",
       "22327   Regulation Of IGF Transport And Uptake By IGFB...      -0.957895   \n",
       "22326   Post-translational Protein Phosphorylation R-H...      -0.957895   \n",
       "27385   Regulation Of IGF Transport And Uptake By IGFB...       0.957042   \n",
       "27384   Post-translational Protein Phosphorylation R-H...       0.957042   \n",
       "68476   Regulation Of IGF Transport And Uptake By IGFB...      -0.956472   \n",
       "68477   Post-translational Protein Phosphorylation R-H...      -0.956472   \n",
       "143780  Detoxification Of Reactive Oxygen Species R-HS...       0.955140   \n",
       "371855                      Synthesis Of PC R-HSA-1483191       0.954797   \n",
       "175058                      Synthesis Of PC R-HSA-1483191      -0.954603   \n",
       "39844   Post-translational Protein Phosphorylation R-H...       0.954468   \n",
       "39843   Regulation Of IGF Transport And Uptake By IGFB...       0.954468   \n",
       "377304          Regulation Of IFNG Signaling R-HSA-877312      -0.954211   \n",
       "319609  Post-translational Protein Phosphorylation R-H...       0.952745   \n",
       "319610  Regulation Of IGF Transport And Uptake By IGFB...       0.952745   \n",
       "323086  Regulation Of IGF Transport And Uptake By IGFB...      -0.951999   \n",
       "323087  Post-translational Protein Phosphorylation R-H...      -0.951999   \n",
       "143474   rRNA Modification In Mitochondrion R-HSA-6793080       0.951905   \n",
       "41113   Cytochrome P450 - Arranged By Substrate Type R...      -0.951814   \n",
       "237219  Regulation Of IGF Transport And Uptake By IGFB...       0.949431   \n",
       "237220  Post-translational Protein Phosphorylation R-H...       0.949431   \n",
       "91242   Post-translational Protein Phosphorylation R-H...       0.948840   \n",
       "91243   Regulation Of IGF Transport And Uptake By IGFB...       0.948840   \n",
       "226405                Viral mRNA Translation R-HSA-192823      -0.947763   \n",
       "57659   Regulation Of IGF Transport And Uptake By IGFB...       0.946816   \n",
       "57658   Post-translational Protein Phosphorylation R-H...       0.946816   \n",
       "363425  IL-6-type Cytokine Receptor Ligand Interaction...       0.946068   \n",
       "319500  Regulation Of IGF Transport And Uptake By IGFB...       0.945640   \n",
       "319499  Post-translational Protein Phosphorylation R-H...       0.945640   \n",
       "110989  Regulation Of IGF Transport And Uptake By IGFB...       0.944881   \n",
       "110990  Post-translational Protein Phosphorylation R-H...       0.944881   \n",
       "282261          Receptor Mediated Mitophagy R-HSA-8934903      -0.943841   \n",
       "375673  Regulation Of IGF Transport And Uptake By IGFB...      -0.943403   \n",
       "375672  Post-translational Protein Phosphorylation R-H...      -0.943403   \n",
       "367555  Cytochrome P450 - Arranged By Substrate Type R...       0.941947   \n",
       "\n",
       "        nes score   p value  shuffled  \n",
       "94309    2.749145  0.005975     False  \n",
       "94308    2.749145  0.005975     False  \n",
       "28586    2.868438  0.004125     False  \n",
       "28587    2.868438  0.004125     False  \n",
       "174148  -2.913465  0.003574     False  \n",
       "174149  -2.913465  0.003574     False  \n",
       "176386  -3.084574  0.002038     False  \n",
       "176387  -3.084574  0.002038     False  \n",
       "222392   3.166789  0.001541     False  \n",
       "56865    2.771649  0.005577     False  \n",
       "56864    2.771649  0.005577     False  \n",
       "158480  -2.808619  0.004975     False  \n",
       "233377   2.732804  0.006280     False  \n",
       "233376   2.732804  0.006280     False  \n",
       "230340  -3.091930  0.001989     False  \n",
       "61856   -2.924281  0.003453     False  \n",
       "22327   -2.850276  0.004368     False  \n",
       "22326   -2.850276  0.004368     False  \n",
       "27385    2.705392  0.006822     False  \n",
       "27384    2.705392  0.006822     False  \n",
       "68476   -2.838184  0.004537     False  \n",
       "68477   -2.838184  0.004537     False  \n",
       "143780   3.238024  0.001204     False  \n",
       "371855   2.862941  0.004197     False  \n",
       "175058  -2.827130  0.004697     False  \n",
       "39844    2.770181  0.005603     False  \n",
       "39843    2.770181  0.005603     False  \n",
       "377304  -2.816713  0.004852     False  \n",
       "319609   2.791679  0.005244     False  \n",
       "319610   2.791679  0.005244     False  \n",
       "323086  -2.788009  0.005303     False  \n",
       "323087  -2.788009  0.005303     False  \n",
       "143474   3.251298  0.001149     False  \n",
       "41113   -2.536189  0.011207     False  \n",
       "237219   2.729118  0.006350     False  \n",
       "237220   2.729118  0.006350     False  \n",
       "91242    2.847893  0.004401     False  \n",
       "91243    2.847893  0.004401     False  \n",
       "226405  -3.035439  0.002402     False  \n",
       "57659    2.750558  0.005949     False  \n",
       "57658    2.750558  0.005949     False  \n",
       "363425   2.706974  0.006790     False  \n",
       "319500   2.733541  0.006266     False  \n",
       "319499   2.733541  0.006266     False  \n",
       "110989   2.899886  0.003733     False  \n",
       "110990   2.899886  0.003733     False  \n",
       "282261  -2.766957  0.005658     False  \n",
       "375673  -2.779569  0.005443     False  \n",
       "375672  -2.779569  0.005443     False  \n",
       "367555   3.012471  0.002591     False  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_results_df.sort_values(by='gsea es score', key=abs, ascending = False).head(50)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gene_dependency_representations",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
